<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Tools for Assessing Clustering • ClustAssess</title>
<script src="lightswitch.js"></script><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Tools for Assessing Clustering">
<meta name="description" content="A set of tools for evaluating clustering robustness using proportion of ambiguously clustered pairs (Senbabaoglu et al. (2014) &lt;doi:10.1038/srep06207&gt;), as well as similarity across methods and method stability using element-centric clustering comparison (Gates et al. (2019) &lt;doi:10.1038/s41598-019-44892-y&gt;). Additionally, this package enables stability-based parameter assessment for graph-based clustering pipelines typical in single-cell data analysis.">
<meta property="og:description" content="A set of tools for evaluating clustering robustness using proportion of ambiguously clustered pairs (Senbabaoglu et al. (2014) &lt;doi:10.1038/srep06207&gt;), as well as similarity across methods and method stability using element-centric clustering comparison (Gates et al. (2019) &lt;doi:10.1038/s41598-019-44892-y&gt;). Additionally, this package enables stability-based parameter assessment for graph-based clustering pipelines typical in single-cell data analysis.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">ClustAssess</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/ClustAssess.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/comparing-soft-and-hierarchical.html">Comparing soft and hierarchical clusterings with element-centric similarity</a></li>
    <li><a class="dropdown-item" href="articles/stability-dim-reduction-flexibility.html">Flexibility of the dimensionality reduction assessment</a></li>
    <li><a class="dropdown-item" href="articles/stability-pipeline-description.html">Stability-based parameter assessment</a></li>
    <li><a class="dropdown-item" href="articles/stability-pipeline-shiny.html">Automatic ClustAssess pipeline. Generating the ClustAssess shiny-app</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="tutorials/index.html">ClustAssess Shiny app Tutorial</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Core-Bioinformatics/ClustAssess/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="clustassess-tools-for-assessing-clustering">ClustAssess: Tools for Assessing Clustering<a class="anchor" aria-label="anchor" href="#clustassess-tools-for-assessing-clustering"></a>
</h1></div>
<p><a href="https://github.com/r-hub/cranlogs.app" class="external-link"><img src="https://cranlogs.r-pkg.org/badges/ClustAssess" alt="rstudio mirror downloads"></a> <a href="https://github.com/r-hub/cranlogs.app" class="external-link"><img src="https://cranlogs.r-pkg.org/badges/grand-total/ClustAssess" alt="rstudio mirror downloads"></a> <a href="https://github.com/Core-Bioinformatics/ClustAssess" class="external-link"><img src="https://img.shields.io/github/languages/code-size/Core-Bioinformatics/ClustAssess.svg"></a> <a href="https://CRAN.R-project.org/package=ClustAssess" class="external-link"><img src="https://www.r-pkg.org/badges/version/ClustAssess?color=orange" alt="CRAN status"></a> <a href="https://github.com/Core-Bioinformatics/ClustAssess" class="external-link"><img src="https://img.shields.io/badge/devel%20version-1.0.0-green.svg"></a> <img src="https://img.shields.io/github/last-commit/Core-Bioinformatics/ClustAssess/main?style=flat&amp;color=white" alt="GitHub last commit (branch)"></p>
<p>This repo contains the ClustAssess R package, which provides a set of tools for evaluating clustering robustness.</p>
<p>Please check the <a href="https://github.com/Core-Bioinformatics/ClustAssessPy" class="external-link">ClustAssessPy</a> module for a Python implementation of the ClustAssess package.</p>
<ul>
<li>
<a href="#the-tools">The Tools</a><ul>
<li><a href="#proportion-of-ambiguously-clustered-pairs-pac">Proportion of Ambiguously Clustered Pairs (PAC)</a></li>
<li><a href="#element-centric-clustering-similarity-ecs">Element Centric Clustering Similarity (ECS)</a></li>
<li><a href="#marker-gene-overlap">Marker Gene Overlap</a></li>
<li><a href="#stability-based-parameter-assessment">Stability-Based Parameter Assessment</a></li>
<li><a href="#interactive-visualization-of-the-stability-assessment-and-downstream-analysis">Interactive Visualization of the Stability Assessment and Downstream Analysis</a></li>
</ul>
</li>
<li><a href="#handling-large-datasets">Handling Large Datasets</a></li>
<li><a href="#installation">Installation</a></li>
<li><a href="#citing-clustassess">Citing ClustAssess</a></li>
<li><a href="#references">References</a></li>
</ul>
</div>
<div class="section level1">
<h1 id="the-tools">The Tools<a class="anchor" aria-label="anchor" href="#the-tools"></a>
</h1>
<p><img src="https://raw.githubusercontent.com/Core-Bioinformatics/ClustAssess/main/man/figures/ClustAssess_diagram_all.png" width="80%/"></p>
<div class="section level2">
<h2 id="proportion-of-ambiguously-clustered-pairs-pac">Proportion of Ambiguously Clustered Pairs (PAC)<a class="anchor" aria-label="anchor" href="#proportion-of-ambiguously-clustered-pairs-pac"></a>
</h2>
<p>To assess clustering robustness, the proportion of ambiguously clustered pairs (PAC) [1] uses a consensus clustering. The rate of element co-clustering is recorded across various numbers of clusters, <em>k</em>. The lower the PAC, the stabler the clustering for that <em>k</em>.</p>
<p>ClustAssess uses a heirarchical clustering as base for the consensus clustering, and an optimized Rcpp [2] implementation to compute the PAC values. To calculate PAC, we write:</p>
<p><code>cc_res = consensus_cluster(your_data, n_reps=50, k_max=20, p_sample=0.8, p_feature=0.8)</code></p>
<p>It is important that the PAC has converged before using it to assess your data; the <code>pac_convergence</code> function can be used to visualize the PAC curves across iterations:</p>
<p><img src="https://raw.githubusercontent.com/Core-Bioinformatics/ClustAssess/main/docs/articles/ClustAssess_files/figure-html/pac-1.png" width="80%/"></p>
<p>As the curves have evened out, we surmize that PAC has converged in this case. If the PAC has not converged, increase the <code>n_reps</code> value.</p>
<p>If the dataset contains &gt;1000 elements, we recommend calculating a geometric sketch [3] of your data of size &lt;1000, and running PAC on that sketch.</p>
<p>A local minimum in the PAC landscape, as visualized below using the <code>pac_landscape</code> function, can be interpreted as an optimal <em>k</em> for the dataset:</p>
<p><img src="https://raw.githubusercontent.com/Core-Bioinformatics/ClustAssess/main/docs/articles/ClustAssess_files/figure-html/pac-2.png" width="80%/"></p>
</div>
<div class="section level2">
<h2 id="element-centric-clustering-similarity-ecs">Element Centric Clustering Similarity (ECS)<a class="anchor" aria-label="anchor" href="#element-centric-clustering-similarity-ecs"></a>
</h2>
<p>Element-centric similarity (ECS) [4] is a per-observation measure of clustering similarity. Briefly, ECS is obtained by constructing a cluster-induced element graph, and subsequently calculating the stationary probability distribution of a random walk with restarts (also known as a personalized Pagerank) over the elements.</p>
<p>In addition to providing per-element information on clustering agreement, ECS avoids several pitfalls associated with other measures of clustering similarity (refer to [4] for more detail).</p>
<p>To compare two clusterings with ECS, we use the <code>element_sim_elscore</code> function:</p>
<p><code>ecs = element_sim_elscore(clustering_result1, clustering_result2, alpha=0.9)</code></p>
<p>where <code>1-alpha</code> is the restart probability of the random walk. We can subsequently visualize the ECS on a PCA of the data:</p>
<p><img src="https://raw.githubusercontent.com/Core-Bioinformatics/ClustAssess/main/docs/articles/comparing-soft-and-hierarchical_files/figure-html/ecs-2.png" width="80%/"></p>
<p>In addition to flat disjoint clusterings (like the result of k-means for example), ClustAssess can also compare overlapping clusterings and hierarchical clusterings; refer to the <code>comparing-soft-and-hierarchical</code> vignette for more detail.</p>
<p>Besides comparing two clusterings with <code>element_sim_elscore</code>, ClustAssess also enables computing the per-element consistency (aka frustration) between a set of clusterings with the <code>element_consistency</code> function, and comparing a set of clusterings with a ground truth partitioning of elements with the <code>element_agreement</code> function.</p>
</div>
<div class="section level2">
<h2 id="marker-gene-overlap">Marker Gene Overlap<a class="anchor" aria-label="anchor" href="#marker-gene-overlap"></a>
</h2>
<p>In the analysis of single-cell RNA-seq data, cluster marker genes are typically used to infer the cell type of clusters. However, two different clustering results may lead to two different sets of markers for the cells. The <code>marker_overlap</code> function allows the comparison of two sets of markers. The output can be either as the number of common marker genes per cell, or as Jaccard similarity (size of intersect divided by size of union) per cell.</p>
<p><img src="https://raw.githubusercontent.com/Core-Bioinformatics/ClustAssess/main/docs/articles/ClustAssess_files/figure-html/jsi-1.png" width="80%/"></p>
</div>
<div class="section level2">
<h2 id="stability-based-parameter-assessment">Stability-Based Parameter Assessment<a class="anchor" aria-label="anchor" href="#stability-based-parameter-assessment"></a>
</h2>
<p>The most common clustering pipeline for single-cell data consists of constructing a nearest-neighbor graph of the cells, followed by community detection to obtain the clustering. This pipeline is available in multiple single-cell toolkits, including Seurat, Monocle v3, and SCANPY. ClustAssess provides several methods to assess the stability of parameter choices that influence the final clustering, for example to evaluate feature sets, we use <code>get_feature_stability_object</code>, and plot the results with <code>plot_feature_stability_boxplot</code>:</p>
<p><img src="https://raw.githubusercontent.com/Core-Bioinformatics/ClustAssess/main/docs/articles/stability-pipeline-description_files/figure-html/stab_boxplot-1.png" width="80%/"></p>
<p>where the higher element-centric consistency (ECC) indicates more stable clustering results across random seeds. For more details, please see <a href="https://core-bioinformatics.github.io/ClustAssess/articles/stability-pipeline-description.html">this vignette</a>.</p>
</div>
<div class="section level2">
<h2 id="interactive-visualization-of-the-stability-assessment-and-downstream-analysis">Interactive Visualization of the Stability Assessment and Downstream Analysis<a class="anchor" aria-label="anchor" href="#interactive-visualization-of-the-stability-assessment-and-downstream-analysis"></a>
</h2>
<p><img src="https://raw.githubusercontent.com/Core-Bioinformatics/ClustAssess/main/man/figures/ClustAssess_app_supp.png" height="40%/"></p>
<p>Another feature of the package consists in providing a Shiny application where the user can interactively visualize the stability assessment of the clustering and perform some downstream analysis steps on the dataset without prior R knwoledge. The application is divided into three major components: - Stability Assessments: the user is presented with the assessment of the clustering stability for each of the three steps of the PhenoGraph pipeline. This component is structured in three tabs, one for each step. The user is prompted to select the stable configurations in order to proceed to the next section. - Comparison of configuration and downstream analysis: the user can visually assess the clustering results and the pattern on expression based on the variety of visualisations provided in this section, which contains UMAP plots (A - one or multiple gene expression, B - metadata distribution), violin and boxplots (G), gene expression pseudobulk heatmap and bubbleplots (E), metadata association heatmaps (C). The biological interpretation is enhanced by the possibility of identifying the marker genes for each cluster (F) and the enrichment analysis of the detected differentially expressed genes (H). The app allows custom annotation of the clusters and the already existing metadata (D). - Sandbox: this tab provides more flexibility for the user in temrs of the selected configuration. This approach is available, but not encouraged, as the results might not be stable and reproducible. This tab also contains elements such as panels for changing the colourscheme of the figures.</p>
<p>The application can be generated using the output of the <code>automatic_stability_assessment</code> function and a normalised expression matrix.</p>
</div>
</div>
<div class="section level1">
<h1 id="handling-large-datasets">Handling Large Datasets<a class="anchor" aria-label="anchor" href="#handling-large-datasets"></a>
</h1>
<p>If your dataset is large, the runtime for the tools described above may be prohibitive. In these cases, we recommend subsampling your data using geometric sketching [3]. In R, the subsampling can be done via reticulate:</p>
<p><code>geosketch &lt;- reticulate::import('geosketch')</code></p>
<p>assuming data.embed contains a dimensionality reduction of your data, you can then call:</p>
<p><code>sketch.indices &lt;- geosketch$gs(data.embed, sketch.size, one_indexed = TRUE)</code></p>
<p>and use the resulting indices for your subsample. For PAC, subsampling to &lt;1000 cells should help, and for ECS and data assessment functions, &lt;5000 cells may be appropriate (and parallelization can further help reduce the runtime).</p>
</div>
<div class="section level1">
<h1 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h1>
<p>ClustAssess can be installed from <a href="https://cran.r-project.org/index.html" class="external-link">CRAN</a>:</p>
<p><code>install.packages("ClustAssess")</code></p>
<p>or from github using remotes:</p>
<p><code>remotes::install_github("Core-Bioinformatics/ClustAssess")</code></p>
<p>The following packages are required for ClustAssess:</p>
<ul>
<li>ComplexHeatmap</li>
<li>dplyr</li>
<li>DT</li>
<li>fastcluster</li>
<li>foreach</li>
<li>glue</li>
<li>Gmedian</li>
<li>ggnewscale</li>
<li>ggplot2</li>
<li>ggrastr</li>
<li>ggrepel</li>
<li>ggtext</li>
<li>gprofiler2</li>
<li>igraph</li>
<li>jsonlite</li>
<li>Matrix (&gt;= 1.5.0)</li>
<li>matrixStats</li>
<li>methods</li>
<li>progress</li>
<li>stringr</li>
<li>paletteer</li>
<li>plotly</li>
<li>qualpalr</li>
<li>RANN</li>
<li>reshape2</li>
<li>rlang</li>
<li>shiny</li>
<li>shinyjs</li>
<li>shinyLP</li>
<li>shinyWidgets</li>
<li>stats</li>
<li>uwot</li>
<li>vioplot</li>
</ul>
<p>To use all stability-based assessment methods, and run all examples and vignettes, the following packages are also needed:</p>
<ul>
<li>colourpicker</li>
<li>dbscan</li>
<li>DelayedMatrixStats</li>
<li>dendextend</li>
<li>devtools</li>
<li>doParallel</li>
<li>doRNG</li>
<li>e1071</li>
<li>knitr</li>
<li>leidenbase</li>
<li>monocle3</li>
<li>patchwork</li>
<li>ragg</li>
<li>reactlog</li>
<li>rhdf5</li>
<li>rmarkdown</li>
<li>RhpcBLASctl</li>
<li>scales</li>
<li>Seurat</li>
<li>SeuratData</li>
<li>SeuratObject</li>
<li>SharedObject</li>
<li>styler</li>
<li>testthat</li>
</ul>
<p>To manually install the required dependencies, please use the following commands:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dplyr"</span>, <span class="st">"DT"</span>, <span class="st">"fastcluster"</span>, <span class="st">"foreach"</span>, <span class="st">"glue"</span>, <span class="st">"Gmedian"</span>, <span class="st">"ggnewscale"</span>, <span class="st">"ggplot2"</span>, <span class="st">"ggrastr"</span>, <span class="st">"ggrepel"</span>, <span class="st">"ggtext"</span>, <span class="st">"gprofiler2"</span>, <span class="st">"igraph"</span>, <span class="st">"jsonlite"</span>, <span class="st">"Matrix"</span>, <span class="st">"matrixStats"</span>, <span class="st">"methods"</span>, <span class="st">"progress"</span>, <span class="st">"stringr"</span>, <span class="st">"paletteer"</span>, <span class="st">"plotly"</span>, <span class="st">"qualpalr"</span>, <span class="st">"RANN"</span>, <span class="st">"reshape2"</span>, <span class="st">"rlang"</span>, <span class="st">"shiny"</span>, <span class="st">"shinyjs"</span>, <span class="st">"shinyLP"</span>, <span class="st">"shinyWidgets"</span>, <span class="st">"stats"</span>, <span class="st">"uwot"</span>, <span class="st">"vioplot"</span><span class="op">)</span>, Ncpus <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ComplexHeatmap"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>To manually install the suggested dependencies, please use the following commands:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"colourpicker"</span>, <span class="st">"dbscan"</span>, <span class="st">"dendextend"</span>, <span class="st">"devtools"</span>, <span class="st">"doParallel"</span>, <span class="st">"doRNG"</span>, <span class="st">"e1071"</span>, <span class="st">"knitr"</span>, <span class="st">"leidenbase"</span>, <span class="st">"patchwork"</span>, <span class="st">"ragg"</span>, <span class="st">"reactlog"</span>, <span class="st">"rmarkdown"</span>, <span class="st">"RhpcBLASctl"</span>, <span class="st">"scales"</span>, <span class="st">"Seurat"</span>, <span class="st">"SeuratObject"</span>, <span class="st">"styler"</span>, <span class="st">"testthat"</span><span class="op">)</span>, Ncpus <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DelayedMatrixStats"</span>, <span class="st">"rhdf5"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'cole-trapnell-lab/monocle3'</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'satijalab/seurat-data'</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"Jiefei-Wang/SharedObject"</span><span class="op">)</span></span></code></pre></div>
<p><em>Note</em>: The package is under active development, and the release branches might contain additional features and bug fixes.</p>
</div>
<div class="section level1">
<h1 id="citing-clustassess">Citing ClustAssess<a class="anchor" aria-label="anchor" href="#citing-clustassess"></a>
</h1>
<p>If you have used ClustAssess in your work, please cite <a href="https://doi.org/10.1101/2022.01.31.478592" class="external-link">Shahsavari et al. 2022</a>:</p>
<pre><code>Shahsavari, A., Munteanu, A., &amp; Mohorianu, I. (2022). ClustAssess: Tools for Assessing the Robustness of Single-Cell Clustering. bioRxiv. https://doi.org/10.1101/2022.01.31.478592</code></pre>
<p>or in BibTex:</p>
<pre><code>@ARTICLE{clustassess,
  title     = "ClustAssess: Tools for Assessing the Robustness of Single-Cell Clustering",
  author    = "Shahsavari, Arash and Munteanu, Andi and Mohorianu, Irina",
  journal   = "bioRxiv",
  year      =  2022,
  url       = "https://doi.org/10.1101/2022.01.31.478592",
  language  = "en"
}</code></pre>
</div>
<div class="section level1">
<h1 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h1>
<p>[1] Șenbabaoğlu, Y., Michailidis, G., &amp; Li, J. Z. (2014). Critical limitations of consensus clustering in class discovery. Scientific reports, 4(1), 1-13. <a href="https://doi.org/10.1038/srep06207" class="external-link uri">https://doi.org/10.1038/srep06207</a></p>
<p>[2] Eddelbuettel, D., &amp; Balamuta, J. J. (2018). Extending R with C++: A brief introduction to Rcpp. The American Statistician, 72(1), 28-36. <a href="https://doi.org/10.1080/00031305.2017.1375990" class="external-link uri">https://doi.org/10.1080/00031305.2017.1375990</a></p>
<p>[3] Hie, B., Cho, H., DeMeo, B., Bryson, B., &amp; Berger, B. (2019). Geometric sketching compactly summarizes the single-cell transcriptomic landscape. Cell systems, 8(6), 483-493. <a href="https://doi.org/10.1016/j.cels.2019.05.003" class="external-link uri">https://doi.org/10.1016/j.cels.2019.05.003</a></p>
<p>[4] Gates, A. J., Wood, I. B., Hetrick, W. P., &amp; Ahn, Y. Y. (2019). Element-centric clustering comparison unifies overlaps and hierarchy. Scientific reports, 9(1), 1-13. <a href="https://doi.org/10.1038/s41598-019-44892-y" class="external-link uri">https://doi.org/10.1038/s41598-019-44892-y</a></p>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=ClustAssess" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/Core-Bioinformatics/ClustAssess/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/Core-Bioinformatics/ClustAssess/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing ClustAssess</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Andi Munteanu <br><small class="roles"> Author, maintainer </small>  </li>
<li>Arash Shahsavari <br><small class="roles"> Author </small>  </li>
<li>Miguel Larraz Lopez de Novales <br><small class="roles"> Author </small>  </li>
<li>Irina Mohorianu <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Andi Munteanu, Arash Shahsavari, Miguel Larraz Lopez de Novales, Irina Mohorianu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
