<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Evaluating single-cell clustering with ClustAssess â€¢ ClustAssess</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Evaluating single-cell clustering with ClustAssess">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ClustAssess</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/ClustAssess.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/comparing-soft-and-hierarchical.html">Comparing soft and hierarchical clusterings with element-centric similarity</a></li>
    <li><a class="dropdown-item" href="../articles/stability-dim-reduction-flexibility.html">Flexibility of the dimensionality reduction assessment</a></li>
    <li><a class="dropdown-item" href="../articles/stability-pipeline-description.html">Stability-based parameter assessment</a></li>
    <li><a class="dropdown-item" href="../articles/stability-pipeline-shiny.html">Automatic ClustAssess pipeline. Generating the ClustAssess shiny-app</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../tutorials/index.html">ClustAssess Shiny app Tutorial</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Core-Bioinformatics/ClustAssess/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<script src="ClustAssess_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Evaluating single-cell clustering with ClustAssess</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Core-Bioinformatics/ClustAssess/blob/HEAD/vignettes/ClustAssess.Rmd" class="external-link"><code>vignettes/ClustAssess.Rmd</code></a></small>
      <div class="d-none name"><code>ClustAssess.Rmd</code></div>
    </div>

    
    
<p>In this vignette we will illustrate several of the tools available in ClustAssess on a small single-cell RNA-seq dataset.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SeuratObject</span></span>
<span><span class="co">#&gt; Loading required package: sp</span></span>
<span><span class="co">#&gt; 'SeuratObject' was built with package 'Matrix' 1.7.0 but the current</span></span>
<span><span class="co">#&gt; version is 1.7.2; it is recomended that you reinstall 'SeuratObject' as</span></span>
<span><span class="co">#&gt; the ABI for 'Matrix' may have changed</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'SeuratObject'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, t</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Core-Bioinformatics/ClustAssess" class="external-link">ClustAssess</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'ggplot2'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     is.element</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/get_theme.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"pbmc_small"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># we will use the pbmc_small single-cell RNA-seq dataset available via Seurat</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmc_small</span>, group.by <span class="op">=</span> <span class="st">"letter.idents"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/setup-1.png" width="672"></p>
<div class="section level2">
<h2 id="proportion-of-ambiguously-clustered-pairs">Proportion of ambiguously clustered pairs<a class="anchor" aria-label="anchor" href="#proportion-of-ambiguously-clustered-pairs"></a>
</h2>
<p>The proportion of ambiguously clustered pairs (PAC) uses consensus clustering to infer the optimal number of clusters for the data, by observing how variably pairs of elements cluster together. The lower the PAC, the more stable the clustering. PAC was originally presented in <a href="https://doi.org/10.1038/srep06207" class="external-link uri">https://doi.org/10.1038/srep06207</a>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># retrieve scaled data for PAC calculation</span></span>
<span><span class="va">pbmc.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, layer <span class="op">=</span> <span class="st">"scale.data"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># perform consensus clustering</span></span>
<span><span class="va">cc.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/consensus_cluster.html">consensus_cluster</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">pbmc.data</span><span class="op">)</span>,</span>
<span>    k_max <span class="op">=</span> <span class="fl">30</span>,</span>
<span>    n_reps <span class="op">=</span> <span class="fl">100</span>,</span>
<span>    p_sample <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>    p_feature <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Calculating consensus clustering</span></span>
<span></span>
<span><span class="co"># assess the PAC convergence for a few values of k - each curve should</span></span>
<span><span class="co"># have converged to some value</span></span>
<span><span class="va">k.plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">6</span>, <span class="fl">8</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/pac_convergence.html">pac_convergence</a></span><span class="op">(</span><span class="va">cc.res</span>, <span class="va">k.plot</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/pac-1.png" width="672"></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># visualize the final PAC across k - there seems to be a local maximum at k=7,</span></span>
<span><span class="co"># indicating that 7 clusters leads to a less stable clustering of the data than</span></span>
<span><span class="co"># nearby values of k</span></span>
<span><span class="fu"><a href="../reference/pac_landscape.html">pac_landscape</a></span><span class="op">(</span><span class="va">cc.res</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/pac-2.png" width="672"></p>
</div>
<div class="section level2">
<h2 id="element-centric-clustering-comparison">Element-centric clustering comparison<a class="anchor" aria-label="anchor" href="#element-centric-clustering-comparison"></a>
</h2>
<p>We compare the similarity of clustering results between methods (in this case, between Louvain community detection and k-means) using Element-Centric Similarity (ECS), which quantifies the clustering similarity per cell. Higher ECS indicates that an observation is clustered similarly across methods. ECS was introduced in <a href="https://doi.org/10.1038/s41598-019-44892-y" class="external-link uri">https://doi.org/10.1038/s41598-019-44892-y</a>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># first, cluster with Louvain algorithm</span></span>
<span><span class="va">pbmc_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">pbmc_small</span>, resolution <span class="op">=</span> <span class="fl">0.8</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmc_small</span>, group.by <span class="op">=</span> <span class="st">"seurat_clusters"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/ecs-1.png" width="672"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># also cluster with PCA+k-means</span></span>
<span><span class="va">pbmc_pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Embeddings.html" class="external-link">Embeddings</a></span><span class="op">(</span><span class="va">pbmc_small</span>, <span class="st">"pca"</span><span class="op">)</span></span>
<span><span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">kmeans_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans</a></span><span class="op">(</span><span class="va">pbmc_pca</span>,</span>
<span>    centers <span class="op">=</span> <span class="fl">3</span>,</span>
<span>    nstart <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    iter.max <span class="op">=</span> <span class="fl">1e3</span></span>
<span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmc_small</span>, group.by <span class="op">=</span> <span class="st">"kmeans_clusters"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/ecs-2.png" width="672"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># where are the clustering results more similar?</span></span>
<span><span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">ecs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/element_sim_elscore.html">element_sim_elscore</a></span><span class="op">(</span></span>
<span>    <span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">seurat_clusters</span>,</span>
<span>    <span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">kmeans_clusters</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmc_small</span>, <span class="st">"ecs"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/ecs-3.png" width="672"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">ecs</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.6632927</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="jaccard-similarity-of-cluster-markers">Jaccard similarity of cluster markers<a class="anchor" aria-label="anchor" href="#jaccard-similarity-of-cluster-markers"></a>
</h2>
<p>As a common step in computational single-cell RNA-seq analyses, discriminative marker genes are identified for each cluster. These markers are then used to infer the cell type of the respective cluster. Here, we compare the markers obtained for each clustering method to ask: how similarly would each cell be interpreted across clustering methods? We compare the markers per cell using the Jaccard similarity (defined as the size of the intersect divided by the size of the overlap) of the marker gene lists. The higher the JSI, the more similar are the marker genes for that cell.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># first, we calculate the markers on the Louvain clustering</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">seurat_clusters</span></span>
<span><span class="va">louvain.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span><span class="va">pbmc_small</span>,</span>
<span>    logfc.threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    min.pct <span class="op">=</span> <span class="fl">0.0</span>,</span>
<span>    test.use <span class="op">=</span> <span class="st">"roc"</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># then we get the markers on the k-means clustering</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">kmeans_clusters</span></span>
<span><span class="va">kmeans.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span><span class="va">pbmc_small</span>,</span>
<span>    logfc.threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    min.pct <span class="op">=</span> <span class="fl">0.0</span>,</span>
<span>    test.use <span class="op">=</span> <span class="st">"roc"</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># next, compare the top 10 markers per cell</span></span>
<span><span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">marker.gene.jsi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/marker_overlap.html">marker_overlap</a></span><span class="op">(</span><span class="va">louvain.markers</span>,</span>
<span>    <span class="va">kmeans.markers</span>,</span>
<span>    <span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">seurat_clusters</span>,</span>
<span>    <span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">kmeans_clusters</span>,</span>
<span>    n <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    rank_by <span class="op">=</span> <span class="st">"power"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># which cells have the same markers, regardless of clustering?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmc_small</span>, <span class="st">"marker.gene.jsi"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/jsi-1.png" width="672"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">marker.gene.jsi</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.5738636</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="element-wise-consistency">Element-wise consistency<a class="anchor" aria-label="anchor" href="#element-wise-consistency"></a>
</h2>
<p>How consistently are cells clustered by k-means? We will rerun k-means clustering 20 times to investigate.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clustering.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">n.reps</span> <span class="op">&lt;-</span> <span class="fl">20</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">n.reps</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># we set nstart=1 and a fairly high iter.max - this should mean that</span></span>
<span>    <span class="co"># the algorithm converges, and that the variability in final clusterings</span></span>
<span>    <span class="co"># depends mainly on the random initial cluster assignments</span></span>
<span>    <span class="va">km.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans</a></span><span class="op">(</span><span class="va">pbmc_pca</span>, centers <span class="op">=</span> <span class="fl">3</span>, nstart <span class="op">=</span> <span class="fl">1</span>, iter.max <span class="op">=</span> <span class="fl">1e3</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span></span>
<span>    <span class="va">clustering.list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">km.res</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># now, we calculate the element-wise consistency (aka frustration), which</span></span>
<span><span class="co"># performs pair-wise comparisons between all 20 clusterings; the</span></span>
<span><span class="co"># consistency is the average per-cell ECS across all comparisons. The higher</span></span>
<span><span class="co"># the consistency, the more consistently is that cell clustered across</span></span>
<span><span class="co"># random seeds.</span></span>
<span><span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">consistency</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/element_consistency.html">element_consistency</a></span><span class="op">(</span><span class="va">clustering.list</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># which cells are clustered more consistently?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmc_small</span>, <span class="st">"consistency"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/frust-1.png" width="672"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">consistency</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.6712922</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.0 (2024-04-24)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Europe/Bucharest</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] ggplot2_3.5.1.9000 ClustAssess_1.0.0  Seurat_5.2.0       SeuratObject_5.0.2</span></span>
<span><span class="co">#&gt; [5] sp_2.1-4          </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RColorBrewer_1.1-3     jsonlite_1.8.9         magrittr_2.0.3        </span></span>
<span><span class="co">#&gt;   [4] spatstat.utils_3.1-2   farver_2.1.2           rmarkdown_2.29        </span></span>
<span><span class="co">#&gt;   [7] fs_1.6.5               ragg_1.3.3             vctrs_0.6.5           </span></span>
<span><span class="co">#&gt;  [10] ROCR_1.0-11            spatstat.explore_3.3-4 htmltools_0.5.8.1     </span></span>
<span><span class="co">#&gt;  [13] progress_1.2.3         sass_0.4.9             sctransform_0.4.1     </span></span>
<span><span class="co">#&gt;  [16] parallelly_1.41.0      KernSmooth_2.23-26     bslib_0.8.0           </span></span>
<span><span class="co">#&gt;  [19] htmlwidgets_1.6.4      desc_1.4.3             ica_1.0-3             </span></span>
<span><span class="co">#&gt;  [22] plyr_1.8.9             plotly_4.10.4          zoo_1.8-12            </span></span>
<span><span class="co">#&gt;  [25] cachem_1.1.0           igraph_2.1.4           mime_0.12             </span></span>
<span><span class="co">#&gt;  [28] lifecycle_1.0.4        iterators_1.0.14       pkgconfig_2.0.3       </span></span>
<span><span class="co">#&gt;  [31] Matrix_1.7-2           R6_2.5.1               fastmap_1.2.0         </span></span>
<span><span class="co">#&gt;  [34] fitdistrplus_1.2-2     future_1.34.0          shiny_1.10.0          </span></span>
<span><span class="co">#&gt;  [37] digest_0.6.37          colorspace_2.1-1       patchwork_1.3.0       </span></span>
<span><span class="co">#&gt;  [40] tensor_1.5             RSpectra_0.16-2        irlba_2.3.5.1         </span></span>
<span><span class="co">#&gt;  [43] textshaping_1.0.0      labeling_0.4.3         progressr_0.15.1      </span></span>
<span><span class="co">#&gt;  [46] spatstat.sparse_3.1-0  httr_1.4.7             polyclip_1.10-7       </span></span>
<span><span class="co">#&gt;  [49] abind_1.4-8            compiler_4.4.0         withr_3.0.2           </span></span>
<span><span class="co">#&gt;  [52] fastDummies_1.7.5      MASS_7.3-64            tools_4.4.0           </span></span>
<span><span class="co">#&gt;  [55] lmtest_0.9-40          httpuv_1.6.15          future.apply_1.11.3   </span></span>
<span><span class="co">#&gt;  [58] goftest_1.2-3          glue_1.8.0             nlme_3.1-166          </span></span>
<span><span class="co">#&gt;  [61] promises_1.3.2         grid_4.4.0             Rtsne_0.17            </span></span>
<span><span class="co">#&gt;  [64] cluster_2.1.8          reshape2_1.4.4         generics_0.1.3        </span></span>
<span><span class="co">#&gt;  [67] gtable_0.3.6           spatstat.data_3.1-4    tidyr_1.3.1           </span></span>
<span><span class="co">#&gt;  [70] hms_1.1.3              data.table_1.16.4      spatstat.geom_3.3-5   </span></span>
<span><span class="co">#&gt;  [73] RcppAnnoy_0.0.22       ggrepel_0.9.6          RANN_2.6.2            </span></span>
<span><span class="co">#&gt;  [76] foreach_1.5.2          pillar_1.10.1          stringr_1.5.1         </span></span>
<span><span class="co">#&gt;  [79] spam_2.11-1            RcppHNSW_0.6.0         later_1.4.1           </span></span>
<span><span class="co">#&gt;  [82] splines_4.4.0          dplyr_1.1.4            lattice_0.22-6        </span></span>
<span><span class="co">#&gt;  [85] survival_3.8-3         deldir_2.0-4           tidyselect_1.2.1      </span></span>
<span><span class="co">#&gt;  [88] miniUI_0.1.1.1         pbapply_1.7-2          knitr_1.49            </span></span>
<span><span class="co">#&gt;  [91] gridExtra_2.3          scattermore_1.2        xfun_0.50             </span></span>
<span><span class="co">#&gt;  [94] matrixStats_1.5.0      stringi_1.8.4          lazyeval_0.2.2        </span></span>
<span><span class="co">#&gt;  [97] yaml_2.3.10            evaluate_1.0.3         codetools_0.2-20      </span></span>
<span><span class="co">#&gt; [100] tibble_3.2.1           cli_3.6.3              uwot_0.2.2            </span></span>
<span><span class="co">#&gt; [103] xtable_1.8-4           reticulate_1.40.0      systemfonts_1.2.1     </span></span>
<span><span class="co">#&gt; [106] munsell_0.5.1          jquerylib_0.1.4        Rcpp_1.0.14           </span></span>
<span><span class="co">#&gt; [109] globals_0.16.3         spatstat.random_3.3-2  png_0.1-8             </span></span>
<span><span class="co">#&gt; [112] fastcluster_1.2.6      spatstat.univar_3.1-1  parallel_4.4.0        </span></span>
<span><span class="co">#&gt; [115] pkgdown_2.1.1          prettyunits_1.2.0      dotCall64_1.2         </span></span>
<span><span class="co">#&gt; [118] listenv_0.9.1          viridisLite_0.4.2      scales_1.3.0          </span></span>
<span><span class="co">#&gt; [121] ggridges_0.5.6         purrr_1.0.2            crayon_1.5.3          </span></span>
<span><span class="co">#&gt; [124] rlang_1.1.5            cowplot_1.1.3</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Andi Munteanu, Arash Shahsavari, Miguel Larraz Lopez de Novales, Irina Mohorianu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
