<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Evaluating single-cell clustering with ClustAssess • ClustAssess</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Evaluating single-cell clustering with ClustAssess">
<meta property="og:description" content="ClustAssess">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ClustAssess</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/ClustAssess.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/comparing-soft-and-hierarchical.html">Comparing soft and hierarchical clusterings with element-centric similarity</a>
    </li>
    <li>
      <a href="../articles/dim_reduction_flexibility.html">Flexibility of the dimensionality reduction assessment</a>
    </li>
    <li>
      <a href="../articles/stability-pipeline-description.html">Stability-based parameter assessment</a>
    </li>
    <li>
      <a href="../articles/stability-pipeline-shiny.html">Automatic ClustAssess pipeline. Generating the ClustAssess shiny-app</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Core-Bioinformatics/ClustAssess/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Evaluating single-cell clustering with
ClustAssess</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Core-Bioinformatics/ClustAssess/blob/HEAD/vignettes/ClustAssess.Rmd" class="external-link"><code>vignettes/ClustAssess.Rmd</code></a></small>
      <div class="hidden name"><code>ClustAssess.Rmd</code></div>

    </div>

    
    
<p>In this vignette we will illustrate several of the tools available in
ClustAssess on a small single-cell RNA-seq dataset.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SeuratObject</span></span>
<span><span class="co">#&gt; Loading required package: sp</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'SeuratObject'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Core-Bioinformatics/ClustAssess" class="external-link">ClustAssess</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"pbmc_small"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># we will use the pbmc_small single-cell RNA-seq dataset available via Seurat</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmc_small</span>, group.by <span class="op">=</span> <span class="st">"letter.idents"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/setup-1.png" width="672"></p>
<div class="section level2">
<h2 id="proportion-of-ambiguously-clustered-pairs">Proportion of ambiguously clustered pairs<a class="anchor" aria-label="anchor" href="#proportion-of-ambiguously-clustered-pairs"></a>
</h2>
<p>The proportion of ambiguously clustered pairs (PAC) uses consensus
clustering to infer the optimal number of clusters for the data, by
observing how variably pairs of elements cluster together. The lower the
PAC, the more stable the clustering. PAC was originally presented in <a href="https://doi.org/10.1038/srep06207" class="external-link uri">https://doi.org/10.1038/srep06207</a>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># retrieve scaled data for PAC calculation</span></span>
<span><span class="va">pbmc.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, slot <span class="op">=</span> <span class="st">"scale.data"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The `slot` argument of `GetAssayData()` is deprecated as of SeuratObject 5.0.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use the `layer` argument instead.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span>
<span></span>
<span><span class="co"># perform consensus clustering</span></span>
<span><span class="va">cc.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/consensus_cluster.html">consensus_cluster</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">pbmc.data</span><span class="op">)</span>,</span>
<span>    k_max <span class="op">=</span> <span class="fl">30</span>,</span>
<span>    n_reps <span class="op">=</span> <span class="fl">100</span>,</span>
<span>    p_sample <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>    p_feature <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Calculating consensus clustering</span></span>
<span></span>
<span><span class="co"># assess the PAC convergence for a few values of k - each curve should</span></span>
<span><span class="co"># have converged to some value</span></span>
<span><span class="va">k.plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">6</span>, <span class="fl">8</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/pac_convergence.html">pac_convergence</a></span><span class="op">(</span><span class="va">cc.res</span>, <span class="va">k.plot</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/pac-1.png" width="672"></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># visualize the final PAC across k - there seems to be a local maximum at k=7,</span></span>
<span><span class="co"># indicating that 7 clusters leads to a less stable clustering of the data than</span></span>
<span><span class="co"># nearby values of k</span></span>
<span><span class="fu"><a href="../reference/pac_landscape.html">pac_landscape</a></span><span class="op">(</span><span class="va">cc.res</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/pac-2.png" width="672"></p>
</div>
<div class="section level2">
<h2 id="element-centric-clustering-comparison">Element-centric clustering comparison<a class="anchor" aria-label="anchor" href="#element-centric-clustering-comparison"></a>
</h2>
<p>We compare the similarity of clustering results between methods (in
this case, between Louvain community detection and k-means) using
element-centric similarity, which quantifies the clustering similarity
per cell. Higher ECS indicates that an observation is clustered
similarly across methods. ECS was introduced in <a href="https://doi.org/10.1038/s41598-019-44892-y" class="external-link uri">https://doi.org/10.1038/s41598-019-44892-y</a>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># first, cluster with Louvain algorithm</span></span>
<span><span class="va">pbmc_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">pbmc_small</span>, resolution <span class="op">=</span> <span class="fl">0.8</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmc_small</span>, group.by <span class="op">=</span> <span class="st">"seurat_clusters"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/ecs-1.png" width="672"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># also cluster with PCA+k-means</span></span>
<span><span class="va">pbmc_pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Embeddings.html" class="external-link">Embeddings</a></span><span class="op">(</span><span class="va">pbmc_small</span>, <span class="st">"pca"</span><span class="op">)</span></span>
<span><span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">kmeans_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans</a></span><span class="op">(</span><span class="va">pbmc_pca</span>,</span>
<span>    centers <span class="op">=</span> <span class="fl">3</span>,</span>
<span>    nstart <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    iter.max <span class="op">=</span> <span class="fl">1e3</span></span>
<span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmc_small</span>, group.by <span class="op">=</span> <span class="st">"kmeans_clusters"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/ecs-2.png" width="672"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># where are the clustering results more similar?</span></span>
<span><span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">ecs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/element_sim_elscore.html">element_sim_elscore</a></span><span class="op">(</span></span>
<span>    <span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">seurat_clusters</span>,</span>
<span>    <span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">kmeans_clusters</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmc_small</span>, <span class="st">"ecs"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/ecs-3.png" width="672"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">ecs</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.6632927</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="jaccard-similarity-of-cluster-markers">Jaccard similarity of cluster markers<a class="anchor" aria-label="anchor" href="#jaccard-similarity-of-cluster-markers"></a>
</h2>
<p>As a common step in computational single-cell RNA-seq analyses,
discriminative marker genes are identified for each cluster. These
markers are then used to infer the cell type of the respective cluster.
Here, we compare the markers obtained for each clustering method to ask:
how similarly would each cell be interpreted across clustering methods?
We compare the markers per cell using the Jaccard similarity (defined as
the size of the intersect divided by the size of the overlap) of the
marker gene lists. The higher the JSI, the more similar are the marker
genes for that cell.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># first, we calculate the markers on the Louvain clustering</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">seurat_clusters</span></span>
<span><span class="va">louvain.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span><span class="va">pbmc_small</span>,</span>
<span>    logfc.threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    min.pct <span class="op">=</span> <span class="fl">0.0</span>,</span>
<span>    test.use <span class="op">=</span> <span class="st">"roc"</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># then we get the markers on the k-means clustering</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">kmeans_clusters</span></span>
<span><span class="va">kmeans.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span><span class="va">pbmc_small</span>,</span>
<span>    logfc.threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    min.pct <span class="op">=</span> <span class="fl">0.0</span>,</span>
<span>    test.use <span class="op">=</span> <span class="st">"roc"</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># next, compare the top 10 markers per cell</span></span>
<span><span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">marker.gene.jsi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/marker_overlap.html">marker_overlap</a></span><span class="op">(</span><span class="va">louvain.markers</span>,</span>
<span>    <span class="va">kmeans.markers</span>,</span>
<span>    <span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">seurat_clusters</span>,</span>
<span>    <span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">kmeans_clusters</span>,</span>
<span>    n <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    rank_by <span class="op">=</span> <span class="st">"power"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># which cells have the same markers, regardless of clustering?</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmc_small</span>, <span class="st">"marker.gene.jsi"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/jsi-1.png" width="672"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">marker.gene.jsi</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.5738636</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="element-wise-consistency">Element-wise consistency<a class="anchor" aria-label="anchor" href="#element-wise-consistency"></a>
</h2>
<p>How consistently are cells clustered by k-means? We will rerun
k-means clustering 20 times to investigate.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clustering.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">n.reps</span> <span class="op">&lt;-</span> <span class="fl">20</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">n.reps</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># we set nstart=1 and a fairly high iter.max - this should mean that</span></span>
<span>    <span class="co"># the algorithm converges, and that the variability in final clusterings</span></span>
<span>    <span class="co"># depends mainly on the random initial cluster assignments</span></span>
<span>    <span class="va">km.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans</a></span><span class="op">(</span><span class="va">pbmc_pca</span>, centers <span class="op">=</span> <span class="fl">3</span>, nstart <span class="op">=</span> <span class="fl">1</span>, iter.max <span class="op">=</span> <span class="fl">1e3</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span></span>
<span>    <span class="va">clustering.list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">km.res</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># now, we calculate the element-wise consistency (aka frustration), which</span></span>
<span><span class="co"># performs pair-wise comparisons between all 20 clusterings; the</span></span>
<span><span class="co"># consistency is the average per-cell ECS across all comparisons. The higher</span></span>
<span><span class="co"># the consistency, the more consistently is that cell clustered across</span></span>
<span><span class="co"># random seeds.</span></span>
<span><span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">consistency</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/element_consistency.html">element_consistency</a></span><span class="op">(</span><span class="va">clustering.list</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># which cells are clustered more consistently?</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmc_small</span>, <span class="st">"consistency"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/frust-1.png" width="672"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">consistency</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.6712922</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.3 (2024-02-29 ucrt)</span></span>
<span><span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Windows 11 x64 (build 22631)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] LC_COLLATE=Romanian_Romania.utf8  LC_CTYPE=Romanian_Romania.utf8   </span></span>
<span><span class="co">#&gt; [3] LC_MONETARY=Romanian_Romania.utf8 LC_NUMERIC=C                     </span></span>
<span><span class="co">#&gt; [5] LC_TIME=Romanian_Romania.utf8    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Europe/Bucharest</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] ClustAssess_1.0.0  Seurat_5.0.3       SeuratObject_5.0.1 sp_2.1-3          </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RColorBrewer_1.1-3     jsonlite_1.8.8         magrittr_2.0.3        </span></span>
<span><span class="co">#&gt;   [4] spatstat.utils_3.0-4   farver_2.1.1           rmarkdown_2.26        </span></span>
<span><span class="co">#&gt;   [7] fs_1.6.3               ragg_1.3.0             vctrs_0.6.5           </span></span>
<span><span class="co">#&gt;  [10] ROCR_1.0-11            memoise_2.0.1          spatstat.explore_3.2-7</span></span>
<span><span class="co">#&gt;  [13] progress_1.2.3         htmltools_0.5.8        sass_0.4.9            </span></span>
<span><span class="co">#&gt;  [16] sctransform_0.4.1      parallelly_1.37.1      KernSmooth_2.23-22    </span></span>
<span><span class="co">#&gt;  [19] bslib_0.7.0            htmlwidgets_1.6.4      desc_1.4.3            </span></span>
<span><span class="co">#&gt;  [22] ica_1.0-3              plyr_1.8.9             plotly_4.10.4         </span></span>
<span><span class="co">#&gt;  [25] zoo_1.8-12             cachem_1.0.8           igraph_2.0.3          </span></span>
<span><span class="co">#&gt;  [28] mime_0.12              lifecycle_1.0.4        iterators_1.0.14      </span></span>
<span><span class="co">#&gt;  [31] pkgconfig_2.0.3        Matrix_1.6-5           R6_2.5.1              </span></span>
<span><span class="co">#&gt;  [34] fastmap_1.1.1          fitdistrplus_1.1-11    future_1.33.2         </span></span>
<span><span class="co">#&gt;  [37] shiny_1.8.1.1          digest_0.6.35          colorspace_2.1-0      </span></span>
<span><span class="co">#&gt;  [40] patchwork_1.2.0        tensor_1.5             RSpectra_0.16-1       </span></span>
<span><span class="co">#&gt;  [43] irlba_2.3.5.1          textshaping_0.3.7      labeling_0.4.3        </span></span>
<span><span class="co">#&gt;  [46] progressr_0.14.0       fansi_1.0.6            spatstat.sparse_3.0-3 </span></span>
<span><span class="co">#&gt;  [49] httr_1.4.7             polyclip_1.10-6        abind_1.4-5           </span></span>
<span><span class="co">#&gt;  [52] compiler_4.3.3         withr_3.0.0            fastDummies_1.7.3     </span></span>
<span><span class="co">#&gt;  [55] highr_0.10             MASS_7.3-60.0.1        tools_4.3.3           </span></span>
<span><span class="co">#&gt;  [58] lmtest_0.9-40          httpuv_1.6.15          future.apply_1.11.2   </span></span>
<span><span class="co">#&gt;  [61] goftest_1.2-3          glue_1.7.0             nlme_3.1-164          </span></span>
<span><span class="co">#&gt;  [64] promises_1.2.1         grid_4.3.3             Rtsne_0.17            </span></span>
<span><span class="co">#&gt;  [67] cluster_2.1.6          reshape2_1.4.4         generics_0.1.3        </span></span>
<span><span class="co">#&gt;  [70] gtable_0.3.4           spatstat.data_3.0-4    tidyr_1.3.1           </span></span>
<span><span class="co">#&gt;  [73] hms_1.1.3              data.table_1.15.4      utf8_1.2.4            </span></span>
<span><span class="co">#&gt;  [76] spatstat.geom_3.2-9    RcppAnnoy_0.0.22       ggrepel_0.9.5         </span></span>
<span><span class="co">#&gt;  [79] RANN_2.6.1             foreach_1.5.2          pillar_1.9.0          </span></span>
<span><span class="co">#&gt;  [82] stringr_1.5.1          spam_2.10-0            RcppHNSW_0.6.0        </span></span>
<span><span class="co">#&gt;  [85] later_1.3.2            splines_4.3.3          dplyr_1.1.4           </span></span>
<span><span class="co">#&gt;  [88] lattice_0.22-5         survival_3.5-8         deldir_2.0-4          </span></span>
<span><span class="co">#&gt;  [91] tidyselect_1.2.1       miniUI_0.1.1.1         pbapply_1.7-2         </span></span>
<span><span class="co">#&gt;  [94] knitr_1.45             gridExtra_2.3          scattermore_1.2       </span></span>
<span><span class="co">#&gt;  [97] xfun_0.43              matrixStats_1.2.0      stringi_1.8.3         </span></span>
<span><span class="co">#&gt; [100] lazyeval_0.2.2         yaml_2.3.8             evaluate_0.23         </span></span>
<span><span class="co">#&gt; [103] codetools_0.2-19       tibble_3.2.1           cli_3.6.2             </span></span>
<span><span class="co">#&gt; [106] uwot_0.1.16.9000       xtable_1.8-4           reticulate_1.35.0     </span></span>
<span><span class="co">#&gt; [109] systemfonts_1.0.6      munsell_0.5.1          jquerylib_0.1.4       </span></span>
<span><span class="co">#&gt; [112] Rcpp_1.0.12            globals_0.16.3         spatstat.random_3.2-3 </span></span>
<span><span class="co">#&gt; [115] png_0.1-8              fastcluster_1.2.6      parallel_4.3.3        </span></span>
<span><span class="co">#&gt; [118] pkgdown_2.0.7          ggplot2_3.5.0          prettyunits_1.2.0     </span></span>
<span><span class="co">#&gt; [121] dotCall64_1.1-1        listenv_0.9.1          viridisLite_0.4.2     </span></span>
<span><span class="co">#&gt; [124] scales_1.3.0           ggridges_0.5.6         crayon_1.5.2          </span></span>
<span><span class="co">#&gt; [127] leiden_0.4.3.1         purrr_1.0.2            rlang_1.1.3           </span></span>
<span><span class="co">#&gt; [130] cowplot_1.1.3</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Arash Shahsavari, Andi Munteanu, Miguel Larraz Lopez de Novales, Irina Mohorianu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
