<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Stability-based parameter assessment • ClustAssess</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Stability-based parameter assessment">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ClustAssess</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/ClustAssess.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/comparing-soft-and-hierarchical.html">Comparing soft and hierarchical clusterings with element-centric similarity</a></li>
    <li><a class="dropdown-item" href="../articles/stability-dim-reduction-flexibility.html">Flexibility of the dimensionality reduction assessment</a></li>
    <li><a class="dropdown-item" href="../articles/stability-pipeline-description.html">Stability-based parameter assessment</a></li>
    <li><a class="dropdown-item" href="../articles/stability-pipeline-shiny.html">Automatic ClustAssess pipeline. Generating the ClustAssess shiny-app</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
<li class="nav-item"><a class="nav-link" href="../tutorials/index.html">ClustAssess Shiny app Tutorial</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Core-Bioinformatics/ClustAssess/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<script src="stability-pipeline-description_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Stability-based parameter assessment</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Core-Bioinformatics/ClustAssess/blob/HEAD/vignettes/stability-pipeline-description.Rmd" class="external-link"><code>vignettes/stability-pipeline-description.Rmd</code></a></small>
      <div class="d-none name"><code>stability-pipeline-description.Rmd</code></div>
    </div>

    
    
<p>Make sure you have Rtools installed: <a href="https://cran.r-project.org/bin/windows/Rtools/rtools42/rtools.html" class="external-link uri">https://cran.r-project.org/bin/windows/Rtools/rtools42/rtools.html</a> In case some packages cannot be installed because of the locks, remove and reinstall them.</p>
<p>Load the libraries.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://prs.ism.ac.jp/~nakama/Rhpc/" class="external-link">RhpcBLASctl</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.satijalab.org/seurat" class="external-link">SeuratData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SeuratObject</span></span>
<span><span class="co">#&gt; Loading required package: sp</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'SeuratObject'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, t</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Core-Bioinformatics/ClustAssess" class="external-link">ClustAssess</a></span><span class="op">)</span></span>
<span><span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/py_install.html" class="external-link">py_install</a></span><span class="op">(</span><span class="st">"leidenalg"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Using virtual environment '/home/andi/.virtualenvs/r-reticulate' ...</span></span>
<span><span class="co">#&gt; + /home/andi/.virtualenvs/r-reticulate/bin/python -m pip install --upgrade --no-user leidenalg</span></span>
<span><span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/py_install.html" class="external-link">py_install</a></span><span class="op">(</span><span class="st">"six"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Using virtual environment '/home/andi/.virtualenvs/r-reticulate' ...</span></span>
<span><span class="co">#&gt; + /home/andi/.virtualenvs/r-reticulate/bin/python -m pip install --upgrade --no-user six</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/packageDescription.html" class="external-link">packageVersion</a></span><span class="op">(</span><span class="st">"ClustAssess"</span><span class="op">)</span> <span class="co"># should be 1.0.0</span></span>
<span><span class="co">#&gt; [1] '1.1.0'</span></span></code></pre></div>
<div class="section level2">
<h2 id="calculating-ecs-ecc">Calculating ECS, ECC<a class="anchor" aria-label="anchor" href="#calculating-ecs-ecc"></a>
</h2>
<p>In the following we will test the functionality of the <code>ClustAssess</code> package in terms of calculating the ECS and ECC.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mb1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">mb2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># calculate the ECS score for each observation</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="../reference/element_sim_elscore.html">element_sim_elscore</a></span><span class="op">(</span><span class="va">mb1</span>, <span class="va">mb2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.6666667 0.2500000 0.6666667 0.7500000 0.7500000 0.7500000 1.0000000</span></span>
<span><span class="co">#&gt; [8] 1.0000000</span></span>
<span><span class="co"># calculate the average ECS</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu"><a href="../reference/element_sim_elscore.html">element_sim_elscore</a></span><span class="op">(</span><span class="va">mb1</span>, <span class="va">mb2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.7291667</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="../reference/element_sim.html">element_sim</a></span><span class="op">(</span><span class="va">mb1</span>, <span class="va">mb2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.7291667</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create a list of 10 random partitions</span></span>
<span><span class="va">list_mbs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample.int</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">5</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">list_mbs</span><span class="op">)</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1] 3 3 3 3 4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; [1] 3 4 3 2 2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt; [1] 4 4 4 4 4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[4]]</span></span>
<span><span class="co">#&gt; [1] 1 1 3 2 3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[5]]</span></span>
<span><span class="co">#&gt; [1] 2 2 4 3 4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[6]]</span></span>
<span><span class="co">#&gt; [1] 1 4 2 3 2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[7]]</span></span>
<span><span class="co">#&gt; [1] 2 3 4 1 2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[8]]</span></span>
<span><span class="co">#&gt; [1] 4 1 4 2 4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[9]]</span></span>
<span><span class="co">#&gt; [1] 3 2 2 2 4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[10]]</span></span>
<span><span class="co">#&gt; [1] 1 2 4 3 1</span></span>
<span><span class="fu"><a href="../reference/element_consistency.html">element_consistency</a></span><span class="op">(</span><span class="va">list_mbs</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.4918519 0.5451852 0.4925926 0.5740741 0.5385185</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="stability-pipeline">Stability pipeline<a class="anchor" aria-label="anchor" href="#stability-pipeline"></a>
</h2>
<div class="section level3">
<h3 id="reading-the-data">Reading the data<a class="anchor" aria-label="anchor" href="#reading-the-data"></a>
</h3>
<p>In this vignette we illustrate a data-driven pipeline for the assessment of optimal clustering parameters for a single-cell RNA-seq dataset. Popular pipelines for single-cell clustering (including <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a>, <a href="https://scanpy.readthedocs.io/en/stable/" class="external-link">SCANPY</a>, and <a href="https://cole-trapnell-lab.github.io/monocle3/" class="external-link">Monocle v3</a>) consist of a sequence of steps, including normalization, dimensionality reduction, nearest-neighbor graph construction, and community detection on that graph. The obtained communities are then the clusters, which can be used for downstream analysis by looking at marker genes, trajectories, and more. ClustAssess offers a set of tools to assess the stability of parameters in the clustering pipeline, for the steps of dimensionality reduction, nearest-neighbor graph construction, and community detection.</p>
<p>To illustrate these tools, we will use the PBMC 3k dataset as it is provided in the <a href="https://github.com/satijalab/seurat-data" class="external-link">SeuratData</a> package. The input is provided as a 13714x 2700 expression level matrix containing the raw quantification (in csv format).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/InstallData.html" class="external-link">InstallData</a></span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The following packages are already installed and will not be</span></span>
<span><span class="co">#&gt; reinstalled: pbmc3k</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/UpdateSeuratObject.html" class="external-link">UpdateSeuratObject</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span></span>
<span><span class="co">#&gt; Validating object structure</span></span>
<span><span class="co">#&gt; Updating object slots</span></span>
<span><span class="co">#&gt; Ensuring keys are in the proper structure</span></span>
<span><span class="co">#&gt; Warning: Assay RNA changing from Assay to Assay</span></span>
<span><span class="co">#&gt; Ensuring keys are in the proper structure</span></span>
<span><span class="co">#&gt; Ensuring feature names don't have underscores or pipes</span></span>
<span><span class="co">#&gt; Updating slots in RNA</span></span>
<span><span class="co">#&gt; Validating object structure for Assay 'RNA'</span></span>
<span><span class="co">#&gt; Object representation is consistent with the most current Seurat version</span></span>
<span><span class="va">pbmc3k</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 13714 features across 2700 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (13714 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html" class="external-link">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">pbmc3k</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span>, col.name <span class="op">=</span> <span class="st">"percent.mito"</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html" class="external-link">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">pbmc3k</span>, pattern <span class="op">=</span> <span class="st">"^RP[SL][[:digit:]]"</span>, col.name <span class="op">=</span> <span class="st">"percent.rp"</span><span class="op">)</span></span></code></pre></div>
<p>The analysis is based on the Seurat pipeline (<a href="https://doi.org/10.1016/j.cell.2021.04.048" class="external-link">Hao et al 2021, Integrated analysis of multimodal single-cell data</a>); after creating the object and recording the mitochondrial and ribosomal gene percentages, those genes are excluded from the downstream analysis.</p>
</div>
<div class="section level3">
<h3 id="quality-checks">Quality checks<a class="anchor" aria-label="anchor" href="#quality-checks"></a>
</h3>
<p>We remove the mithocondrial and ribosomal genes.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remove MT and RP genes</span></span>
<span><span class="va">all.index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">MT.index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"^MT-"</span>, x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">RP.index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"^RP[SL][[:digit:]]"</span>, x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="va">pbmc3k</span><span class="op">[</span><span class="op">!</span><span class="op">(</span><span class="op">(</span><span class="va">all.index</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">MT.index</span><span class="op">)</span> <span class="op">|</span> <span class="op">(</span><span class="va">all.index</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">RP.index</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span></code></pre></div>
<p>We plot the violin plots to visualise the distribution of the number of features, number of counts, percentage of mitochondrial genes, and percentage of ribosomal genes.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc3k</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"percent.mito"</span>, <span class="st">"percent.rp"</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="stability-pipeline-description_files/figure-html/qc_violin-1.png" width="672"></p>
<p>After the quality control step, we filter the dataset using the following criteria: - the number of features is less than 2000 - the number of counts is less than 2500 - the percentage of mitochondrial genes is less than 7% - the percentage of ribosomal genes is greater than 7%</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">pbmc3k</span>, <span class="va">nFeature_RNA</span> <span class="op">&lt;</span> <span class="fl">2000</span> <span class="op">&amp;</span> <span class="va">nCount_RNA</span> <span class="op">&lt;</span> <span class="fl">2500</span> <span class="op">&amp;</span> <span class="va">percent.mito</span> <span class="op">&lt;</span> <span class="fl">7</span> <span class="op">&amp;</span> <span class="va">percent.rp</span> <span class="op">&gt;</span> <span class="fl">7</span><span class="op">)</span></span></code></pre></div>
<p>We plot again the violin plots to see the distribution of the features after the filtering step.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc3k</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"percent.mito"</span>, <span class="st">"percent.rp"</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="stability-pipeline-description_files/figure-html/qc_violin_post-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="normalization-and-scaling">Normalization and scaling<a class="anchor" aria-label="anchor" href="#normalization-and-scaling"></a>
</h3>
<p>The data is normalized and scaled using the Seurat’s <code>NormalizeData</code> and <code>ScaleData</code> methods. The variable feature set is obtained by running the <code>FindVariableFeatures</code> function from the same package.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">pbmc3k</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">pbmc3k</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>, nfeatures <span class="op">=</span> <span class="fl">3000</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/dimnames.html" class="external-link">dimnames</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">var_features</span> <span class="op">&lt;-</span> <span class="va">pbmc3k</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">var.features</span></span>
<span><span class="va">n_abundant</span> <span class="op">&lt;-</span> <span class="fl">3000</span></span>
<span><span class="va">most_abundant_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">)</span>,</span>
<span>    decreasing <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">pbmc3k</span>, features <span class="op">=</span> <span class="va">features</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">pbmc3k</span>,</span>
<span>    npcs <span class="op">=</span> <span class="fl">30</span>,</span>
<span>    approx <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    features <span class="op">=</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">most_abundant_genes</span>, <span class="va">pbmc3k</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">var.features</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">pbmc3k</span>,</span>
<span>    reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>    dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>,</span>
<span>    n.neighbors <span class="op">=</span> <span class="fl">30</span>,</span>
<span>    min.dist <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>    metric <span class="op">=</span> <span class="st">"cosine"</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">raw_umap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">umap</span><span class="op">@</span><span class="va">cell.embeddings</span><span class="op">)</span></span></code></pre></div>
<p>We generate the following plots: A - the distribution of the predefined Seurat celltype annotations; B – E distributions of sequencing depths (B), number of features (C), and percentages of reads incident to mitochondrial genes (D) and ribosomal genes (E) illustrated using the colour gradient.</p>
<p><img src="stability-pipeline-description_files/figure-html/qc-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="effect-of-the-random-seed">Effect of the random seed<a class="anchor" aria-label="anchor" href="#effect-of-the-random-seed"></a>
</h3>
<p>To illustrate the necessity of obtaining reproducible and robust clustering outputs, we will perform only the clustering step several times by only changing the random seed.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing nearest neighbor graph</span></span>
<span><span class="co">#&gt; Computing SNN</span></span>
<span><span class="va">clustering_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">1</span>, by <span class="op">=</span> <span class="fl">100</span>, length.out <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">pbmc3k</span>, algorithm <span class="op">=</span> <span class="fl">3</span>, random.seed <span class="op">=</span> <span class="va">i</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">$</span><span class="va">seurat_clusters</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>We assess the stability of the clustering by calculating the consistency of the 10 clustering outputs.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ecc_value</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/merge_partitions.html">merge_partitions</a></span><span class="op">(</span><span class="va">clustering_list</span>, return_ecs_matrix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/fivenum.html" class="external-link">fivenum</a></span><span class="op">(</span><span class="va">ecc_value</span><span class="op">$</span><span class="va">ecc</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.5437092 0.9953488 1.0000000 1.0000000 1.0000000</span></span></code></pre></div>
<p>The similarity between the distinct partitions can be assessed by accessing the <code>ecs_matrix</code> field.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">ecc_value</span><span class="op">$</span><span class="va">ecs_matrix</span><span class="op">)</span></span>
<span><span class="co">#&gt;           [,1]      [,2]      [,3]</span></span>
<span><span class="co">#&gt; [1,] 1.0000000 0.9915337 0.9950592</span></span>
<span><span class="co">#&gt; [2,] 0.9915337 1.0000000 0.9866320</span></span>
<span><span class="co">#&gt; [3,] 0.9950592 0.9866320 1.0000000</span></span></code></pre></div>
<p>The difference between clusters can be visualised using the ECC distribution. Also, comparing the output of the clustering using the second and the fourth seed, we notice that some cells from the top-left island change their identity between the fifth and seventh cluster.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gplots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">gplots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">raw_umap</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">umap_1</span>, y <span class="op">=</span> <span class="va">umap_2</span>, colour <span class="op">=</span> <span class="va">ecc_value</span><span class="op">$</span><span class="va">ecc</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">raw_umap</span><span class="op">$</span><span class="va">clustering</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">clustering_list</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">gplots</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">raw_umap</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">umap_1</span>, y <span class="op">=</span> <span class="va">umap_2</span>, colour <span class="op">=</span> <span class="va">clustering</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">raw_umap</span><span class="op">$</span><span class="va">clustering</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">clustering_list</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">gplots</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">raw_umap</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">umap_1</span>, y <span class="op">=</span> <span class="va">umap_2</span>, colour <span class="op">=</span> <span class="va">clustering</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">gplots</span><span class="op">)</span></span></code></pre></div>
<p><img src="stability-pipeline-description_files/figure-html/phenograph_different_plots-1.png" width="1440"></p>
</div>
<div class="section level3">
<h3 id="applying-clustassess">Applying ClustAssess<a class="anchor" aria-label="anchor" href="#applying-clustassess"></a>
</h3>
<p><em>Note</em>: The following sections provide an example of applying the ClustAssess summary assessment methods. The values of the parameters should be varied and critically assessed for each individual dataset. Also, we recommend increasing the number of repetitions used for stability inference, to ensure reliable and robust conclusions.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_repetitions</span> <span class="op">&lt;-</span> <span class="fl">10</span></span></code></pre></div>
<p>The following analysis is performed by varying the seed multiple times. The pipeline can be run in parallel, so by initialising a backend with multiple number of cores, the task will be split into the processess. Setting the value of 1 will result in a sequential run of all iterations. We encourage the usage of a higher number of cores when possible, as it will decrease the runtime of the assessment pipeline. Please note that the number of cores should be chosen carefully, as each newly created process will add to the memory usage.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">RhpcBLASctl</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RhpcBLASctl/man/RhpcBLASctl-package.html" class="external-link">blas_set_num_threads</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">ncores</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">ncores</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">my_cluster</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster</a></span><span class="op">(</span></span>
<span>        <span class="va">ncores</span>,</span>
<span>        type <span class="op">=</span> <span class="st">"PSOCK"</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span>    <span class="fu">doParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/doParallel/man/registerDoParallel.html" class="external-link">registerDoParallel</a></span><span class="op">(</span>cl <span class="op">=</span> <span class="va">my_cluster</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="section level4">
<h4 id="dimensionality-reduction">Dimensionality reduction<a class="anchor" aria-label="anchor" href="#dimensionality-reduction"></a>
</h4>
<p>The first parameter that we will evaluate and note to substantially influence the dimensionality reduction (linear or non-linear) is the feature set. The Seurat pipeline relies on the highly variable genes (default) for the computation of the Principal Components. The top x most variable genes or most abundant genes can be selected, where x varies according to the characteristics of the dataset; frequently used values are 500, 1000, 2000, 3000 etc.</p>
<p>To showcase the importance of this parameter, we assess three different sets of features:</p>
<ul>
<li>
<strong>highly variable genes</strong>: this set can be obtained in the Seurat pipeline by using either the <code>SCTransform</code> or <code>FindVariableFeatures</code> method. Depending on the method, using the default parameters will get a number of 2000 or 3000 genes.</li>
<li>
<strong>most abundant genes</strong>: this set can be obtained by sorting the genes based on their expression level.</li>
<li>
<strong>intersection between highly variable and most abundant genes</strong>: this set is obtained by intersecting a subset of the most abundant genes and the highly variable feature set. Taking the intersection will lead to a smaller number of genes compared to the initial sets.</li>
</ul>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">steps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">500</span>, to <span class="op">=</span> <span class="fl">3000</span>, by <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="va">ma_hv_genes_intersection_sets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">steps</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">most_abundant_genes</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">]</span>, <span class="va">var_features</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ma_hv_genes_intersection</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="va">union</span>, <span class="va">ma_hv_genes_intersection_sets</span><span class="op">)</span></span>
<span><span class="va">ma_hv_steps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">ma_hv_genes_intersection_sets</span>, <span class="va">length</span><span class="op">)</span></span></code></pre></div>
<p><code>assess_feature_stability</code> is a function that explores the stability of a feature set. The input consists of a normalized expression matrix, that is summarized using either PCA or UMAP. The obtained embedding is then processed using one of the community detection methods that are used in the <code>Seurat</code> package. This process is iterated across random seeds. The method outputs a list containing an UMAP embedding, the cell assignment corresponding to the most frequent partition and the Element-Centric Consistency (ECC) across all communities obtained throughout the iterations. The higher the Element-Centric Consistency, the more stable the clustering. To increase the robustness of the stability assessment, we also vary the resolution parameter that directly impacts the number of clusters.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca_feature_stability_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="va">c</span>,</span>
<span>    <span class="fu"><a href="../reference/assess_feature_stability.html">assess_feature_stability</a></span><span class="op">(</span></span>
<span>        data_matrix <span class="op">=</span> <span class="va">pbmc3k</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">scale.data</span>,</span>
<span>        feature_set <span class="op">=</span> <span class="va">most_abundant_genes</span>,</span>
<span>        resolution <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">0.1</span>, to <span class="op">=</span> <span class="fl">1</span>, by <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>,</span>
<span>        steps <span class="op">=</span> <span class="va">steps</span>,</span>
<span>        n_repetitions <span class="op">=</span> <span class="va">n_repetitions</span>,</span>
<span>        feature_type <span class="op">=</span> <span class="st">"MA"</span>,</span>
<span>        graph_reduction_type <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>        umap_arguments <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>            min_dist <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>            n_neighbors <span class="op">=</span> <span class="fl">30</span>,</span>
<span>            metric <span class="op">=</span> <span class="st">"cosine"</span></span>
<span>        <span class="op">)</span>,</span>
<span>        ecs_thresh <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        clustering_algorithm <span class="op">=</span> <span class="fl">1</span></span>
<span>    <span class="op">)</span>,</span>
<span>    <span class="fu"><a href="../reference/assess_feature_stability.html">assess_feature_stability</a></span><span class="op">(</span></span>
<span>        data_matrix <span class="op">=</span> <span class="va">pbmc3k</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">scale.data</span>,</span>
<span>        feature_set <span class="op">=</span> <span class="va">var_features</span>,</span>
<span>        resolution <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">0.1</span>, to <span class="op">=</span> <span class="fl">1</span>, by <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>,</span>
<span>        steps <span class="op">=</span> <span class="va">steps</span>,</span>
<span>        n_repetitions <span class="op">=</span> <span class="va">n_repetitions</span>,</span>
<span>        feature_type <span class="op">=</span> <span class="st">"HV"</span>,</span>
<span>        graph_reduction_type <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>        umap_arguments <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>            min_dist <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>            n_neighbors <span class="op">=</span> <span class="fl">30</span>,</span>
<span>            metric <span class="op">=</span> <span class="st">"cosine"</span></span>
<span>        <span class="op">)</span>,</span>
<span>        ecs_thresh <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        clustering_algorithm <span class="op">=</span> <span class="fl">1</span></span>
<span>    <span class="op">)</span>,</span>
<span>    <span class="fu"><a href="../reference/assess_feature_stability.html">assess_feature_stability</a></span><span class="op">(</span></span>
<span>        data_matrix <span class="op">=</span> <span class="va">pbmc3k</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">scale.data</span>,</span>
<span>        feature_set <span class="op">=</span> <span class="va">ma_hv_genes_intersection</span>,</span>
<span>        steps <span class="op">=</span> <span class="va">ma_hv_steps</span>,</span>
<span>        resolution <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">0.1</span>, to <span class="op">=</span> <span class="fl">1</span>, by <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>,</span>
<span>        n_repetitions <span class="op">=</span> <span class="va">n_repetitions</span>,</span>
<span>        feature_type <span class="op">=</span> <span class="st">"MA_HV"</span>,</span>
<span>        graph_reduction_type <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>        umap_arguments <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>            min_dist <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>            n_neighbors <span class="op">=</span> <span class="fl">30</span>,</span>
<span>            metric <span class="op">=</span> <span class="st">"cosine"</span></span>
<span>        <span class="op">)</span>,</span>
<span>        ecs_thresh <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        clustering_algorithm <span class="op">=</span> <span class="fl">1</span></span>
<span>    <span class="op">)</span>,</span>
<span>    SIMPLIFY <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The function <code>assess_feature_stability</code> outputs the stability of different feature sets and concatenates the results to ease cross-comparisons. The plot illustrates the stability assessment, across 30 runs, each with different random seeds, on the three gene sets described above. We assessed the stability of each set on incremental number of selected genes. Thus, the boxplots are arranged in groups of three (the number of evaluated feature sets); above each boxplot we specify the number of elements per subset. The user can interogate the stability values that are obtained for each resolution value, as indicated in the left-side panel, or the overall stability assessment (which is determined by collecting the median of the ECC-distribution of each resolution value), as it can be seen in the right-side panel.</p>
<p>For the PBMC case study, a conclusion that is supported based on this plot is that configurations such as MA 500, 2000, 2500, 3000, HV 500, 1000, 1500, 2500 or MA_HV 65, 175, 425 are suitable choices for the Principal Components Analysis, as the ECC score is high with a low-variance distribution.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/plot_feature_per_resolution_stability_boxplot.html">plot_feature_per_resolution_stability_boxplot</a></span><span class="op">(</span><span class="va">pca_feature_stability_object</span>, resolution <span class="op">=</span> <span class="fl">0.8</span>, text_size <span class="op">=</span> <span class="fl">2.5</span>, boxplot_width <span class="op">=</span> <span class="fl">0.4</span>, dodge_width <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>        legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>        legend.justification <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/plot_feature_overall_stability_boxplot.html">plot_feature_overall_stability_boxplot</a></span><span class="op">(</span><span class="va">pca_feature_stability_object</span>, text_size <span class="op">=</span> <span class="fl">2.5</span>, boxplot_width <span class="op">=</span> <span class="fl">0.4</span>, dodge_width <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>        legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>        legend.justification <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: A numeric `legend.position` argument in `theme()` was deprecated in ggplot2</span></span>
<span><span class="co">#&gt; 3.5.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use the `legend.position.inside` argument of `theme()` instead.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span></code></pre></div>
<p><img src="stability-pipeline-description_files/figure-html/stab_boxplot-1.png" width="1344"></p>
<p>Another angle for assessing the stability is centered on the comparison between consecutive steps, for each feature set, performed using the Element-Centric Similarity on the most frequent partitions from each step. The aim is evaluating the effect of increasing the number of genes on the final partitions, and, indirectly, determining the transition from the signal to the noise zone. For the PBMC case study, we observe an increase in similarity between consecutive steps for the MA and MA_HV genes, suggesting that selecting more genes would lead to a more robust partitioning.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span></span>
<span><span class="fu"><a href="../reference/plot_feature_per_resolution_stability_incremental.html">plot_feature_per_resolution_stability_incremental</a></span><span class="op">(</span><span class="va">pca_feature_stability_object</span>, dodge_width <span class="op">=</span> <span class="fl">0.6</span>, resolution <span class="op">=</span> <span class="fl">0.8</span>, text_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>        legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>        legend.justification <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span><span class="fu"><a href="../reference/plot_feature_overall_stability_incremental.html">plot_feature_overall_stability_incremental</a></span><span class="op">(</span><span class="va">pca_feature_stability_object</span>, dodge_width <span class="op">=</span> <span class="fl">0.6</span>, text_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>        legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>        legend.justification <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="stability-pipeline-description_files/figure-html/stab_ecs_inc-1.png" width="1344"></p>
<p>To enhance the summarized information presented as boxplots, we use <code>plot_feature_stability_mb_facet</code> to visualize the corresponding facet plot displaying the most frequent partitions for each feature set and step on an UMAP embedding. This plot can reveal additional insights i.e. the stability of a gene set might be explained in part by the scattered distribution of the cells across multiple islands. For the PBMC case study we notice a similar topology for all three feature sets.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_feature_stability_mb_facet.html">plot_feature_stability_mb_facet</a></span><span class="op">(</span></span>
<span>    feature_object_list <span class="op">=</span> <span class="va">pca_feature_stability_object</span>,</span>
<span>    resolution <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>    text_size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>    n_facet_cols <span class="op">=</span> <span class="fl">6</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="stability-pipeline-description_files/figure-html/stab_mb-1.png" width="1440"></p>
<p>To further understand the areas of instability, we provide an additional plot where we display, on the same UMAPs as before, the Element-Centric Consistency score of the partitions obtained for each step and feature set. In our case, we notice that the source of instability is located in the main island of cells. These visualizations can help us understand the relationship between stability and the topology of the data.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_feature_stability_ecs_facet.html">plot_feature_stability_ecs_facet</a></span><span class="op">(</span><span class="va">pca_feature_stability_object</span>, resolution <span class="op">=</span> <span class="fl">0.8</span>, n_facet_cols <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span></code></pre></div>
<p><img src="stability-pipeline-description_files/figure-html/stab_ecs-1.png" width="1440"></p>
<p>For the PBMC case study, based on the plots presented above, we conclude that the feature set does not highly impact the topology of the data. Judging by the consistency across multiple runs with different random seeds and the incremental stability, we have select the top 1000 Highly Variable genes.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="va">pbmc3k</span><span class="op">[</span><span class="va">var_features</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span><span class="op">]</span>, <span class="op">]</span></span>
<span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">pbmc3k</span>,</span>
<span>    npcs <span class="op">=</span> <span class="fl">30</span>,</span>
<span>    approx <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    features <span class="op">=</span> <span class="va">var_features</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span><span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">pbmc3k</span>,</span>
<span>    reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>    dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>,</span>
<span>    n.neighbors <span class="op">=</span> <span class="fl">30</span>,</span>
<span>    min.dist <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>    metric <span class="op">=</span> <span class="st">"cosine"</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="graph-construction">Graph construction<a class="anchor" aria-label="anchor" href="#graph-construction"></a>
</h4>
<p>The next step in a standard single-cell analysis pipeline is building the graph using the nearest neighbour algorithm. The following parameters influence the final partitioning:</p>
<ul>
<li>
<strong>base embedding</strong>: the graph can be built on either the PCA or the UMAP embedding (using the expression matrix isn’t recommended, as the distances would be noisier and the runtime would increase)</li>
<li><strong>the number of neighbours</strong></li>
<li>
<strong>the graph type</strong>: the graph can be either unweighted (NN case) or based on a weighted Shared-Nearest Neighbours (SNN) graph. For the latter, the weights are computed using the Jaccard Similarity Index (JSI) between the neighbourhoods of two cells.</li>
</ul>
<p><code>get_nn_conn_comps</code> is a method used to link the number of neighbours and the number of connected components (a connected component is a subgraph within which there exists a path between every pair of nodes). The method accepts as input both PCA and UMAP reductions; the distribution of the number of connected components obtainable for the given number of neighbours is calculated across random seeds. The output is a list containing, for each number of neighbours, an array with the number of connected components obtained on different seeds. As before, the objects can be concatenated to facilitate comparisons between different configurations and reductions.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nn_conn_comps_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_nn_conn_comps.html">get_nn_conn_comps</a></span><span class="op">(</span></span>
<span>    embedding <span class="op">=</span> <span class="va">pbmc3k</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">pca</span><span class="op">@</span><span class="va">cell.embeddings</span>,</span>
<span>    n_neigh_sequence <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">5</span>, to <span class="op">=</span> <span class="fl">30</span>, by <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    n_repetitions <span class="op">=</span> <span class="va">n_repetitions</span>,</span>
<span>    include_umap <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    umap_arguments <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        min_dist <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>        n_neighbors <span class="op">=</span> <span class="fl">30</span>,</span>
<span>        metric <span class="op">=</span> <span class="st">"cosine"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The following plot describes the covariation between the number neighbours and the number of connected components obtained using both PCA and UMAP reductions as base for graph building. As the number of neighbours increases, the number of connected components decreases (this is an expected result, as increasing the number of neighbours result in a better connected graph). Please note that the number of connected components provides a lower bound on the number of clusters we can obtain by downstream community detection algorithms such as Louvain and Leiden.</p>
<p>Another comparison may be performed on the PCA and UMAP embeddings to assess graph connectivity. We note that, when using UMAPs, groups of cells are more separated compared to PCAs, for which a quick convergence to a connected graph (which is a graph where there exists a path between every two nodes) is obtained. The link between the number of connected components and the functional interpretation of the results is that the former could be interpreted as the minimum number of clusters that could be obtained with the given number of neighbours e.g. for the UMAP case, to obtain 5 communities from the clustering step, we would require at least 5 nearest neighbours.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_connected_comps_evolution.html">plot_connected_comps_evolution</a></span><span class="op">(</span><span class="va">nn_conn_comps_object</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in ggplot2::scale_y_continuous(breaks = chosen_breaks, trans =</span></span>
<span><span class="co">#&gt; "log10"): <span style="color: #00BB00;">log-10</span> transformation introduced infinite values.</span></span></code></pre></div>
<p><img src="stability-pipeline-description_files/figure-html/comps_evo-1.png" width="672"></p>
<p><code>get_nn_importance</code> is a method used for assessing the impact of number of nearest neighbours, of the graph type (NN or SNN) and of the base embedding (PCA or UMAP) in determining the final number of clusters. The input is either a normalized expression matrix or a PCA embedding, a range for the number of neighbours and the reduction type used for building the graph. The graph type is not provided, as the method will check both cases i.e. NN and SNN. The output comprises a list of configurations (reduction type, and the graph type). Each configuration will contain a list of unique partitions obtained from multiple runs (iterations) for each selected value of the nearest neighbours.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nn_importance_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="va">c</span>,</span>
<span>    <span class="fu"><a href="../reference/assess_nn_stability.html">assess_nn_stability</a></span><span class="op">(</span></span>
<span>        embedding <span class="op">=</span> <span class="va">pbmc3k</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">pca</span><span class="op">@</span><span class="va">cell.embeddings</span>,</span>
<span>        n_neigh_sequence <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">5</span>, to <span class="op">=</span> <span class="fl">30</span>, by <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>,</span>
<span>        n_repetitions <span class="op">=</span> <span class="va">n_repetitions</span>,</span>
<span>        graph_reduction_type <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>        ecs_thresh <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        clustering_algorithm <span class="op">=</span> <span class="fl">1</span></span>
<span>    <span class="op">)</span>,</span>
<span>    <span class="fu"><a href="../reference/assess_nn_stability.html">assess_nn_stability</a></span><span class="op">(</span></span>
<span>        embedding <span class="op">=</span> <span class="va">pbmc3k</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">pca</span><span class="op">@</span><span class="va">cell.embeddings</span>,</span>
<span>        n_neigh_sequence <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">5</span>, to <span class="op">=</span> <span class="fl">30</span>, by <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>,</span>
<span>        n_repetitions <span class="op">=</span> <span class="va">n_repetitions</span>,</span>
<span>        graph_reduction_type <span class="op">=</span> <span class="st">"UMAP"</span>,</span>
<span>        ecs_thresh <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        umap_arguments <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>            min_dist <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>            n_neighbors <span class="op">=</span> <span class="fl">30</span>,</span>
<span>            metric <span class="op">=</span> <span class="st">"cosine"</span></span>
<span>        <span class="op">)</span>,</span>
<span>        clustering_algorithm <span class="op">=</span> <span class="fl">1</span></span>
<span>    <span class="op">)</span>,</span>
<span>    SIMPLIFY <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We input the resulting object into <code>plot_n_neigh_k_correspondence</code>, to generate the following plot; it summarizes the relationship between the number of nearest neighbours and k, the number of clusters. As for the # neighbours vs # connected components correspondence plot, we note a similar descending trend of the number of clusters as the number of nearest neighbours increases. The plot also illustrates the difference between the two graph types: SNN has a tighter distribution of k (over multiple iterations) compared to NN. If the initial object contains graphs based on both UMAP and PCA embedding, this plot will showcase the impact of this choice, as well. For the PBMC dataset, we notice that the UMAP-based graph is partitioned into a greater number of clusters compared to the PCA case.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_n_neigh_k_correspondence.html">plot_n_neigh_k_correspondence</a></span><span class="op">(</span><span class="va">nn_importance_object</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in ggplot2::scale_y_continuous(breaks = chosen_breaks, trans =</span></span>
<span><span class="co">#&gt; "log10"): <span style="color: #00BB00;">log-10</span> transformation introduced infinite values.</span></span></code></pre></div>
<p><img src="stability-pipeline-description_files/figure-html/n_k_corr-1.png" width="960"></p>
<p>The stability of these parameters can be also evaluated using the Element-Centric Consistency applied on the partition list obtained over the multiple runs. The following summary plot underlines the evolution of the consistency as the number of neighbours increases; the PCA graphs yields stable results even for smaller number of nearest neighbours, whereas the UMAP exhibits a stability improvement as the number of neighbours increases. For the same number of nearest neighbors, SNN is more stable than NN.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_n_neigh_ecs.html">plot_n_neigh_ecs</a></span><span class="op">(</span><span class="va">nn_importance_object</span><span class="op">)</span></span></code></pre></div>
<p><img src="stability-pipeline-description_files/figure-html/n_neigh_ecs-1.png" width="960"></p>
</div>
<div class="section level4">
<h4 id="graph-clustering">Graph clustering<a class="anchor" aria-label="anchor" href="#graph-clustering"></a>
</h4>
<p>The final step in a standard single-cell analysis pipeline is applying a graph-based clustering method. Choosing a community detection algorithm has a significant impact on the partitioning results. Using the <code>get_clustering_difference</code> we assess the stability and reproducibility of results obtained using various graph clustering methods available in the Seurat package: Louvain, Louvain refined, SLM and Leiden.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">adj_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">pca</span><span class="op">@</span><span class="va">cell.embeddings</span>, k.param <span class="op">=</span> <span class="fl">15</span>, nn.method <span class="op">=</span> <span class="st">"rann"</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span><span class="op">$</span><span class="va">snn</span></span></code></pre></div>
<p>As previously suggested, the resolution parameter affects the number of clusters obtained. The <code>get_resolution_importance</code> evaluates the effect of a range of values for the resolution parameter on the stability of the output. Besides the resolution, the user can input multiple values of the following parameters: number of neighbours, the graph type and the clustering method. The function will return a list associated with each parameter value combination. For each resolution value, the list will contain multiple sublists of partitions corresponding to specific number of clusters.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clustering_stability</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assess_clustering_stability.html">assess_clustering_stability</a></span><span class="op">(</span></span>
<span>    graph_adjacency_matrix <span class="op">=</span> <span class="va">adj_matrix</span>,</span>
<span>    resolution <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">0.1</span>, to <span class="op">=</span> <span class="fl">2</span>, by <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>,</span>
<span>    n_repetitions <span class="op">=</span> <span class="va">n_repetitions</span>,</span>
<span>    clustering_algorithm <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,</span>
<span>    ecs_thresh <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Unable to set up conda environment r-reticulate</span></span>
<span><span class="co">#&gt; run in terminal:</span></span>
<span><span class="co">#&gt; conda init</span></span>
<span><span class="co">#&gt; conda create -n r-reticulate</span></span>
<span><span class="co">#&gt; conda environment r-reticulate installed</span></span>
<span><span class="co">#&gt; Unable to install python modules igraph and leidenalg</span></span>
<span><span class="co">#&gt; run in terminal:</span></span>
<span><span class="co">#&gt; conda install -n r-reticulate -c conda-forge leidenalg python-igraph pandas umap-learn</span></span>
<span><span class="co">#&gt; python modules igraph and leidenalg installed</span></span></code></pre></div>
<p>The stability of the methods is evaluated using the Element-Centric Consistency (ECC), which is applied on the partition list obtained over multiple runs. The following summary plot underlines the variation of the consistency as the resolution parameter increases. Above each boxplot, we display the number of clusters corresponding to the most frequent partition for the clustering method with the specific resolution parameter. Increasing the resolution leads to more clusters. We notice a high stability for all four methods for a low number of clusters. Starting with <code>k = 8</code>, Leiden and SLM perform better than Louvain and Louvain refined.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="../reference/plot_clustering_per_value_stability.html">plot_clustering_per_value_stability</a></span><span class="op">(</span><span class="va">clustering_stability</span>, <span class="st">"k"</span><span class="op">)</span> <span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>        legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>        legend.justification <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span></span>
<span>      <span class="op">)</span>,</span>
<span>    <span class="fu"><a href="../reference/plot_clustering_per_value_stability.html">plot_clustering_per_value_stability</a></span><span class="op">(</span><span class="va">clustering_stability</span>, <span class="st">"resolution"</span><span class="op">)</span> <span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>        legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>        legend.justification <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span>  <span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="stability-pipeline-description_files/figure-html/unnamed-chunk-9-1.png" width="1632"></p>
<p>The stats can be further summarised using functions such as the mean or the median and the overall displayed can be visualised in the following plot. This utility of this figure is proven on the task of choosing the clustering method that is overall the most stable. In our case, SLM is the most stable clustering method.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/plot_clustering_overall_stability.html">plot_clustering_overall_stability</a></span><span class="op">(</span><span class="va">clustering_stability</span>, <span class="st">"k"</span><span class="op">)</span> <span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>        legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>        legend.justification <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span></span>
<span>      <span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/plot_clustering_overall_stability.html">plot_clustering_overall_stability</a></span><span class="op">(</span><span class="va">clustering_stability</span>, <span class="st">"resolution"</span>, <span class="va">mean</span><span class="op">)</span> <span class="op">+</span> <span class="co"># function(x) { quantile(x, probs = 0.25)[[1]]})</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>        legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>        legend.justification <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="stability-pipeline-description_files/figure-html/unnamed-chunk-10-1.png" width="960"></p>
<p>The resulting object can be visualized with <code>plot_k_resolution_corresp</code> as shown in the code below; we showcase the relationship between the number of clusters and the resolution value. This plot also provides information on suitable resolution values for predefined number of clusters. The colour gradient represents either the frequency of the partitions having <em>k</em> clusters or the ECC of them. It can also be used as proxy to describe the co-variation between (k, resolution). Lighter (higher) values indicate that little variation is observed, on changes of the random seed, between the number of clusters and the resolution value. The size illustrates the frequency of the most common partition when the resolution and the number of clusters values are fixed.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_k_resolution_corresp.html">plot_k_resolution_corresp</a></span><span class="op">(</span><span class="va">clustering_stability</span>, dodge_width <span class="op">=</span> <span class="fl">0.7</span>, summary_function <span class="op">=</span> <span class="va">min</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"resolution - k correspondence with ecs threshold = 1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="stability-pipeline-description_files/figure-html/k_res_corr_1-1.png" width="1920"></p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_k_resolution_corresp.html">plot_k_resolution_corresp</a></span><span class="op">(</span><span class="va">clustering_stability</span>, dodge_width <span class="op">=</span> <span class="fl">0.7</span>, colour_information <span class="op">=</span> <span class="st">"freq_k"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"resolution - k correspondence with ecs threshold = 1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="stability-pipeline-description_files/figure-html/k_res_corr_1-2.png" width="1920"></p>
<p>The following plot showcases the co-variation between the stability of the number of clusters and the number of different partitions resulting from changes on the seed or the resolution parameter. A high number of different partitions indicates a lower stability for a given number of clusters. The colour gradient is proportional to the frequency of most common partition having a fixed number of clusters or their ECC as an indicator of robustness. The size indicates the frequency of the partition having <em>k</em> clusters relative to the total number of runs and should provide additional information whether the behaviour described by the colour is replicated in multiple instances or not. We note that, even if we obtain a high number of different partitions, if the frequency of the most common one is close to 1, then the overall stability is high. Observing a high number of partitions, each with low frequency, indicates high instability. For the SLM clustering, k = 3, 4, 5, 6, 7, 8, 9 are stable choices for the number of clusters that should be used on the downstream analysis.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_k_n_partitions.html">plot_k_n_partitions</a></span><span class="op">(</span><span class="va">clustering_stability</span>, dodge_width <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"k - # partitions correspondence with ecs threshold = 1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="stability-pipeline-description_files/figure-html/k_n_part_1-1.png" width="1440"></p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_k_n_partitions.html">plot_k_n_partitions</a></span><span class="op">(</span><span class="va">clustering_stability</span>, dodge_width <span class="op">=</span> <span class="fl">0.5</span>, colour_information <span class="op">=</span> <span class="st">"freq_part"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"k - # partitions correspondence with ecs threshold = 1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="stability-pipeline-description_files/figure-html/k_n_part_1-2.png" width="1440"></p>
<p>Close the connections opened when using multiple cores.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">foreach</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/foreach/man/registerDoSEQ.html" class="external-link">registerDoSEQ</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>We conclude by summarizing the runtime for executing each step of the assessment pipeline. The results observed on other systems may vary; the runtime is also affected by the size of the date (i.e number of cells and features), the data itself and by actual values for the various parameters.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span></span>
<span>    <span class="st">"Feature stability methods runtime:"</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/format.html" class="external-link">format</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">stop_time_feature_stability</span> <span class="op">-</span> <span class="va">start_time_feature_stability</span>,</span>
<span>        units <span class="op">=</span> <span class="st">"mins"</span></span>
<span>    <span class="op">)</span><span class="op">)</span>, <span class="st">"minutes"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Feature stability methods runtime: 7.33879 minutes"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span></span>
<span>    <span class="st">"NN - # connected components methods runtime:"</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/format.html" class="external-link">format</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">stop_time_nn_conn</span> <span class="op">-</span> <span class="va">start_time_nn_conn</span>,</span>
<span>        units <span class="op">=</span> <span class="st">"mins"</span></span>
<span>    <span class="op">)</span><span class="op">)</span>, <span class="st">"minutes"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "NN - # connected components methods runtime: 0.5985121 minutes"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span></span>
<span>    <span class="st">"NN stability methods runtime:"</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/format.html" class="external-link">format</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">stop_time_nn_importance</span> <span class="op">-</span> <span class="va">start_time_nn_importance</span>,</span>
<span>        units <span class="op">=</span> <span class="st">"mins"</span></span>
<span>    <span class="op">)</span><span class="op">)</span>, <span class="st">"minutes"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "NN stability methods runtime: 2.524402 minutes"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span></span>
<span>    <span class="st">"Clustering stability methods runtime:"</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/format.html" class="external-link">format</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">stop_time_clustering</span> <span class="op">-</span> <span class="va">start_time_clustering</span>,</span>
<span>        units <span class="op">=</span> <span class="st">"mins"</span></span>
<span>    <span class="op">)</span><span class="op">)</span>, <span class="st">"minutes"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Clustering stability methods runtime: 6.496472 minutes"</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Europe/Bucharest</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] leiden_0.4.3.1                ClustAssess_1.1.0            </span></span>
<span><span class="co">#&gt;  [3] ggplot2_3.5.1                 Seurat_5.1.0                 </span></span>
<span><span class="co">#&gt;  [5] SeuratObject_5.0.2            sp_2.1-4                     </span></span>
<span><span class="co">#&gt;  [7] Matrix_1.7-0                  patchwork_1.2.0              </span></span>
<span><span class="co">#&gt;  [9] pbmcMultiome.SeuratData_0.1.4 pbmc3k.SeuratData_3.1.4      </span></span>
<span><span class="co">#&gt; [11] SeuratData_0.2.2.9002         RhpcBLASctl_0.23-42          </span></span>
<span><span class="co">#&gt; [13] devtools_2.4.5                usethis_2.2.3                </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RColorBrewer_1.1-3     jsonlite_1.8.8         magrittr_2.0.3        </span></span>
<span><span class="co">#&gt;   [4] ggbeeswarm_0.7.2       spatstat.utils_3.0-5   farver_2.1.2          </span></span>
<span><span class="co">#&gt;   [7] rmarkdown_2.27         fs_1.6.4               ragg_1.4.0            </span></span>
<span><span class="co">#&gt;  [10] vctrs_0.6.5            ROCR_1.0-11            memoise_2.0.1         </span></span>
<span><span class="co">#&gt;  [13] spatstat.explore_3.3-1 progress_1.2.3         htmltools_0.5.8.1     </span></span>
<span><span class="co">#&gt;  [16] sass_0.4.9             sctransform_0.4.1      parallelly_1.37.1     </span></span>
<span><span class="co">#&gt;  [19] KernSmooth_2.23-20     bslib_0.7.0            htmlwidgets_1.6.4     </span></span>
<span><span class="co">#&gt;  [22] desc_1.4.3             ica_1.0-3              plyr_1.8.9            </span></span>
<span><span class="co">#&gt;  [25] plotly_4.10.4          zoo_1.8-12             cachem_1.1.0          </span></span>
<span><span class="co">#&gt;  [28] igraph_2.0.3           iterators_1.0.14       mime_0.12             </span></span>
<span><span class="co">#&gt;  [31] lifecycle_1.0.4        pkgconfig_2.0.3        R6_2.5.1              </span></span>
<span><span class="co">#&gt;  [34] fastmap_1.2.0          fitdistrplus_1.2-1     future_1.33.2         </span></span>
<span><span class="co">#&gt;  [37] shiny_1.8.1.1          digest_0.6.36          colorspace_2.1-0      </span></span>
<span><span class="co">#&gt;  [40] rprojroot_2.0.4        tensor_1.5             RSpectra_0.16-1       </span></span>
<span><span class="co">#&gt;  [43] irlba_2.3.5.1          pkgload_1.4.0          textshaping_0.4.0     </span></span>
<span><span class="co">#&gt;  [46] labeling_0.4.3         progressr_0.14.0       spatstat.sparse_3.1-0 </span></span>
<span><span class="co">#&gt;  [49] httr_1.4.7             polyclip_1.10-6        abind_1.4-5           </span></span>
<span><span class="co">#&gt;  [52] compiler_4.4.1         here_1.0.1             remotes_2.5.0         </span></span>
<span><span class="co">#&gt;  [55] withr_3.0.0            fastDummies_1.7.3      pkgbuild_1.4.4        </span></span>
<span><span class="co">#&gt;  [58] MASS_7.3-55            rappdirs_0.3.3         sessioninfo_1.2.2     </span></span>
<span><span class="co">#&gt;  [61] tools_4.4.1            vipor_0.4.7            lmtest_0.9-40         </span></span>
<span><span class="co">#&gt;  [64] beeswarm_0.4.0         httpuv_1.6.15          future.apply_1.11.2   </span></span>
<span><span class="co">#&gt;  [67] goftest_1.2-3          glue_1.7.0             nlme_3.1-155          </span></span>
<span><span class="co">#&gt;  [70] promises_1.3.0         grid_4.4.1             Rtsne_0.17            </span></span>
<span><span class="co">#&gt;  [73] cluster_2.1.2          reshape2_1.4.4         generics_0.1.3        </span></span>
<span><span class="co">#&gt;  [76] gtable_0.3.5           spatstat.data_3.1-2    tidyr_1.3.1           </span></span>
<span><span class="co">#&gt;  [79] hms_1.1.3              data.table_1.15.4      BiocGenerics_0.50.0   </span></span>
<span><span class="co">#&gt;  [82] spatstat.geom_3.3-2    RcppAnnoy_0.0.22       foreach_1.5.2         </span></span>
<span><span class="co">#&gt;  [85] ggrepel_0.9.5          RANN_2.6.1             pillar_1.10.1         </span></span>
<span><span class="co">#&gt;  [88] stringr_1.5.1          spam_2.10-0            RcppHNSW_0.6.0        </span></span>
<span><span class="co">#&gt;  [91] later_1.3.2            splines_4.4.1          dplyr_1.1.4           </span></span>
<span><span class="co">#&gt;  [94] lattice_0.22-6         survival_3.2-13        deldir_2.0-4          </span></span>
<span><span class="co">#&gt;  [97] tidyselect_1.2.1       miniUI_0.1.1.1         pbapply_1.7-2         </span></span>
<span><span class="co">#&gt; [100] knitr_1.49             gridExtra_2.3          scattermore_1.2       </span></span>
<span><span class="co">#&gt; [103] xfun_0.50              SharedObject_1.19.1    matrixStats_1.3.0     </span></span>
<span><span class="co">#&gt; [106] stringi_1.8.4          lazyeval_0.2.2         yaml_2.3.9            </span></span>
<span><span class="co">#&gt; [109] evaluate_0.24.0        codetools_0.2-18       tibble_3.2.1          </span></span>
<span><span class="co">#&gt; [112] cli_3.6.3              uwot_0.2.2             xtable_1.8-4          </span></span>
<span><span class="co">#&gt; [115] reticulate_1.38.0      systemfonts_1.1.0      munsell_0.5.1         </span></span>
<span><span class="co">#&gt; [118] jquerylib_0.1.4        Rcpp_1.0.12            globals_0.16.3        </span></span>
<span><span class="co">#&gt; [121] spatstat.random_3.3-1  png_0.1-8              ggrastr_1.0.2         </span></span>
<span><span class="co">#&gt; [124] spatstat.univar_3.0-0  parallel_4.4.1         ellipsis_0.3.2        </span></span>
<span><span class="co">#&gt; [127] pkgdown_2.1.3          prettyunits_1.2.0      dotCall64_1.1-1       </span></span>
<span><span class="co">#&gt; [130] profvis_0.3.8          urlchecker_1.0.1       listenv_0.9.1         </span></span>
<span><span class="co">#&gt; [133] viridisLite_0.4.2      scales_1.3.0           ggridges_0.5.6        </span></span>
<span><span class="co">#&gt; [136] crayon_1.5.3           purrr_1.0.2            rlang_1.1.4           </span></span>
<span><span class="co">#&gt; [139] cowplot_1.1.3</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Andi Munteanu, Arash Shahsavari, Miguel Larraz Lopez de Novales, Irina Mohorianu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
