<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Flexibility of the dimensionality reduction assessment • ClustAssess</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Flexibility of the dimensionality reduction assessment">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ClustAssess</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/ClustAssess.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/comparing-soft-and-hierarchical.html">Comparing soft and hierarchical clusterings with element-centric similarity</a>
    </li>
    <li>
      <a href="../articles/stability-dim-reduction-flexibility.html">Flexibility of the dimensionality reduction assessment</a>
    </li>
    <li>
      <a href="../articles/stability-pipeline-description.html">Stability-based parameter assessment</a>
    </li>
    <li>
      <a href="../articles/stability-pipeline-shiny.html">Automatic ClustAssess pipeline. Generating the ClustAssess shiny-app</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Core-Bioinformatics/ClustAssess/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><script src="stability-dim-reduction-flexibility_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Flexibility of the dimensionality reduction assessment</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Core-Bioinformatics/ClustAssess/blob/HEAD/vignettes/stability-dim-reduction-flexibility.Rmd" class="external-link"><code>vignettes/stability-dim-reduction-flexibility.Rmd</code></a></small>
      <div class="hidden name"><code>stability-dim-reduction-flexibility.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">is_seuratdata</span> <span class="op">&lt;-</span> <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="http://www.satijalab.org/seurat" class="external-link">"SeuratData"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## ── <span style="font-weight: bold;">Installed datasets</span> ──────────────────────────────── SeuratData v0.2.2.9001 ──</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">pbmc3k</span> 3.1.4</span></span></code></pre>
<pre><code><span><span class="co">## ────────────────────────────────────── Key ─────────────────────────────────────</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BB00;">✔</span> Dataset loaded successfully</span></span>
<span><span class="co">## <span style="color: #BBBB00;">❯</span> Dataset built with a newer version of Seurat than installed</span></span>
<span><span class="co">## <span style="color: #BB0000;">❓</span> Unknown version of Seurat installed</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="va">is_seuratdata</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'satijalab/seurat-data'</span>, upgrade <span class="op">=</span> <span class="st">"never"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">is_harmony</span> <span class="op">&lt;-</span> <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/immunogenomics/harmony" class="external-link">"harmony"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="va">is_harmony</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"harmony"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">is_rsamtool</span> <span class="op">&lt;-</span> <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/Rsamtools" class="external-link">"Rsamtools"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'BiocGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     IQR, mad, sd, var, xtabs</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">##     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">##     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">##     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">##     Position, rank, rbind, Reduce, rownames, sapply, setdiff, table,</span></span>
<span><span class="co">##     tapply, union, unique, unsplit, which.max, which.min</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'S4Vectors'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:utils':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     findMatches</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand.grid, I, unname</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'Biostrings'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     strsplit</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="va">is_rsamtool</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"Rsamtools"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: SeuratObject</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: sp</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'sp'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:IRanges':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     %over%</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'SeuratObject'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:Biostrings':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:GenomicRanges':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:GenomeInfoDb':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:IRanges':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:S4Vectors':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:BiocGenerics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect, t</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.satijalab.org/seurat" class="external-link">SeuratData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Core-Bioinformatics/ClustAssess" class="external-link">ClustAssess</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/immunogenomics/harmony" class="external-link">harmony</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'data.table'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:GenomicRanges':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     shift</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:IRanges':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     shift</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:S4Vectors':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     first, second</span></span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_repetitions</span> <span class="op">&lt;-</span> <span class="fl">30</span></span></code></pre></div>
<p>The <code>matrix processing</code> parameter of the <code>assess_feature_stability</code> function is a function that enables the user to specify any method to perform the dimensionality reduction prior to applying the UMAP algorithm and the clustering pipeline. By default, the dimensionality reduction used in <code>ClustAssess</code> is a precise PCA using the <code>prcomp</code> package. However, this function can be easily changed, as it will be shown in the following examples.</p>
<div class="section level2">
<h2 id="clustassess-using-pca">ClustAssess using PCA<a class="anchor" aria-label="anchor" href="#clustassess-using-pca"></a>
</h2>
<p>For the PCA example, we will use the PBMC 3k dataset from the <code>SeuratData</code> package. The preprocessing of the dataset is identical with the one performed in the stability pipeline vignette.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/InstallData.html" class="external-link">InstallData</a></span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: The following packages are already installed and will not be</span></span>
<span><span class="co">## reinstalled: pbmc3k</span></span></code></pre>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/UpdateSeuratObject.html" class="external-link">UpdateSeuratObject</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Validating object structure</span></span></code></pre>
<pre><code><span><span class="co">## Updating object slots</span></span></code></pre>
<pre><code><span><span class="co">## Ensuring keys are in the proper structure</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Assay RNA changing from Assay to Assay</span></span></code></pre>
<pre><code><span><span class="co">## Ensuring keys are in the proper structure</span></span></code></pre>
<pre><code><span><span class="co">## Ensuring feature names don't have underscores or pipes</span></span></code></pre>
<pre><code><span><span class="co">## Updating slots in RNA</span></span></code></pre>
<pre><code><span><span class="co">## Validating object structure for Assay 'RNA'</span></span></code></pre>
<pre><code><span><span class="co">## Object representation is consistent with the most current Seurat version</span></span></code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html" class="external-link">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">pbmc3k</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span>, col.name <span class="op">=</span> <span class="st">"percent.mito"</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html" class="external-link">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">pbmc3k</span>, pattern <span class="op">=</span> <span class="st">"^RP[SL][[:digit:]]"</span>, col.name <span class="op">=</span> <span class="st">"percent.rp"</span><span class="op">)</span></span>
<span><span class="co"># remove MT and RP genes</span></span>
<span><span class="va">all.index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">MT.index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"^MT-"</span>, x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">RP.index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"^RP[SL][[:digit:]]"</span>, x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="va">pbmc3k</span><span class="op">[</span><span class="op">!</span><span class="op">(</span><span class="op">(</span><span class="va">all.index</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">MT.index</span><span class="op">)</span> <span class="op">|</span> <span class="op">(</span><span class="va">all.index</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">RP.index</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">pbmc3k</span>, <span class="va">nFeature_RNA</span> <span class="op">&lt;</span> <span class="fl">2000</span> <span class="op">&amp;</span> <span class="va">nCount_RNA</span> <span class="op">&lt;</span> <span class="fl">2500</span> <span class="op">&amp;</span> <span class="va">percent.mito</span> <span class="op">&lt;</span> <span class="fl">7</span> <span class="op">&amp;</span> <span class="va">percent.rp</span> <span class="op">&gt;</span> <span class="fl">7</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">pbmc3k</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">pbmc3k</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>, nfeatures <span class="op">=</span> <span class="fl">3000</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/dimnames.html" class="external-link">dimnames</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">var_features</span> <span class="op">&lt;-</span> <span class="va">pbmc3k</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">var.features</span></span>
<span><span class="va">n_abundant</span> <span class="op">&lt;-</span> <span class="fl">3000</span></span>
<span><span class="va">most_abundant_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">)</span>,</span>
<span>    decreasing <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">pbmc3k</span>, features <span class="op">=</span> <span class="va">features</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>We notice that the <code>seurat_annotations</code> column has some missing values. For simplicity, we will replace them with “NA”.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mask</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">$</span><span class="va">seurat_annotations</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span><span class="op">$</span><span class="va">seurat_annotations</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">$</span><span class="va">seurat_annotations</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span><span class="op">$</span><span class="va">seurat_annotations</span><span class="op">[</span><span class="va">mask</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"NA"</span></span></code></pre></div>
<p>Select the features used for the stability assessment.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/dimnames.html" class="external-link">dimnames</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">var_features</span> <span class="op">&lt;-</span> <span class="va">pbmc3k</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">var.features</span></span>
<span><span class="va">n_abundant</span> <span class="op">&lt;-</span> <span class="fl">3000</span></span>
<span><span class="va">most_abundant_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">)</span>,</span>
<span>    decreasing <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">steps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">500</span>, to <span class="op">=</span> <span class="fl">3000</span>, by <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="va">ma_hv_genes_intersection_sets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">steps</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">most_abundant_genes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">x</span><span class="op">]</span>, <span class="va">var_features</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">x</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ma_hv_genes_intersection</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="va">union</span>, <span class="va">ma_hv_genes_intersection_sets</span><span class="op">)</span></span>
<span><span class="va">ma_hv_steps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">ma_hv_genes_intersection_sets</span>, <span class="va">length</span><span class="op">)</span></span></code></pre></div>
<p>Assess the stability of the dimensionality reduction when PCA is used as dimensionality reduction.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">matrix_processing_function</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">dt_mtx</span>, <span class="va">actual_npcs</span> <span class="op">=</span> <span class="fl">30</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">actual_npcs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">actual_npcs</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">dt_mtx</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%/%</a></span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span>    <span class="fu">RhpcBLASctl</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RhpcBLASctl/man/RhpcBLASctl-package.html" class="external-link">blas_set_num_threads</a></span><span class="op">(</span><span class="fu">foreach</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/foreach/man/getDoParWorkers.html" class="external-link">getDoParWorkers</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">embedding</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">dt_mtx</span>, rank. <span class="op">=</span> <span class="va">actual_npcs</span><span class="op">)</span><span class="op">$</span><span class="va">x</span></span>
<span></span>
<span>    <span class="fu">RhpcBLASctl</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RhpcBLASctl/man/RhpcBLASctl-package.html" class="external-link">blas_set_num_threads</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">embedding</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dt_mtx</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">embedding</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">actual_npcs</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">embedding</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">pca_feature_stability</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assess_feature_stability.html">assess_feature_stability</a></span><span class="op">(</span></span>
<span>    data_matrix <span class="op">=</span> <span class="va">pbmc3k</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">scale.data</span>,</span>
<span>    feature_set <span class="op">=</span> <span class="va">most_abundant_genes</span>,</span>
<span>    resolution <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">0.1</span>, to <span class="op">=</span> <span class="fl">1</span>, by <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>,</span>
<span>    steps <span class="op">=</span> <span class="va">steps</span>,</span>
<span>    n_repetitions <span class="op">=</span> <span class="va">n_repetitions</span>,</span>
<span>    feature_type <span class="op">=</span> <span class="st">"MA"</span>,</span>
<span>    graph_reduction_type <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>    matrix_processing <span class="op">=</span> <span class="va">matrix_processing_function</span>,</span>
<span>    umap_arguments <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        min_dist <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>        n_neighbors <span class="op">=</span> <span class="fl">30</span>,</span>
<span>        metric <span class="op">=</span> <span class="st">"cosine"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    ecs_thresh <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    clustering_algorithm <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: executing %dopar% sequentially: no parallel backend registered</span></span></code></pre>
<p>Plot the distribution of the celltypes on the UMAP embedding obtained on the top 1000 Most Abundant genes.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">umap_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">pca_feature_stability</span><span class="op">$</span><span class="va">embedding_list</span><span class="op">$</span><span class="va">MA</span><span class="op">$</span><span class="st">"1000"</span><span class="op">)</span></span>
<span><span class="va">umap_df</span><span class="op">$</span><span class="va">celltypes</span> <span class="op">&lt;-</span> <span class="va">pbmc3k</span><span class="op">$</span><span class="va">seurat_annotations</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">umap_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">UMAP_1</span>, y <span class="op">=</span> <span class="va">UMAP_2</span>, color <span class="op">=</span> <span class="va">celltypes</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="stability-dim-reduction-flexibility_files/figure-html/umap_pca-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="clustassess-using-harmony">ClustAssess using Harmony<a class="anchor" aria-label="anchor" href="#clustassess-using-harmony"></a>
</h2>
<p>We can also modify the function by adding an addition post-processing step to the PCA. In this example, we will use the Harmony correction to remove the “batch effect” created by the celltypes. <em>Note</em>: This example is meant to exemplify how to use the Harmony correction in the ClusAssess pipeline. The batch correction is actually not needed in the PBMC 3k dataset.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">matrix_processing_function</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">dt_mtx</span>, <span class="va">actual_npcs</span> <span class="op">=</span> <span class="fl">30</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">actual_npcs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">actual_npcs</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">dt_mtx</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%/%</a></span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span>    <span class="fu">RhpcBLASctl</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RhpcBLASctl/man/RhpcBLASctl-package.html" class="external-link">blas_set_num_threads</a></span><span class="op">(</span><span class="fu">foreach</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/foreach/man/getDoParWorkers.html" class="external-link">getDoParWorkers</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">embedding</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">dt_mtx</span>, rank. <span class="op">=</span> <span class="va">actual_npcs</span><span class="op">)</span><span class="op">$</span><span class="va">x</span></span>
<span></span>
<span>    <span class="fu">RhpcBLASctl</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RhpcBLASctl/man/RhpcBLASctl-package.html" class="external-link">blas_set_num_threads</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">embedding</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dt_mtx</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">embedding</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">actual_npcs</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">embedding</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/harmony/man/RunHarmony.html" class="external-link">RunHarmony</a></span><span class="op">(</span><span class="va">embedding</span>, <span class="va">pbmc3k</span><span class="op">$</span><span class="va">seurat_annotations</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">embedding</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">pca_harmony_feature_stability</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assess_feature_stability.html">assess_feature_stability</a></span><span class="op">(</span></span>
<span>    data_matrix <span class="op">=</span> <span class="va">pbmc3k</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">scale.data</span>,</span>
<span>    feature_set <span class="op">=</span> <span class="va">most_abundant_genes</span>,</span>
<span>    resolution <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">0.1</span>, to <span class="op">=</span> <span class="fl">1</span>, by <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>,</span>
<span>    steps <span class="op">=</span> <span class="va">steps</span>,</span>
<span>    n_repetitions <span class="op">=</span> <span class="va">n_repetitions</span>,</span>
<span>    feature_type <span class="op">=</span> <span class="st">"MA"</span>,</span>
<span>    graph_reduction_type <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>    matrix_processing <span class="op">=</span> <span class="va">matrix_processing_function</span>,</span>
<span>    umap_arguments <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        min_dist <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>        n_neighbors <span class="op">=</span> <span class="fl">30</span>,</span>
<span>        metric <span class="op">=</span> <span class="st">"cosine"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    ecs_thresh <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    clustering_algorithm <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Plot the distribution of the celltypes on the UMAP embedding obtained on the top 1000 Most Abundant genes.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">umap_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">pca_harmony_feature_stability</span><span class="op">$</span><span class="va">embedding_list</span><span class="op">$</span><span class="va">MA</span><span class="op">$</span><span class="st">"1000"</span><span class="op">)</span></span>
<span><span class="va">umap_df</span><span class="op">$</span><span class="va">celltypes</span> <span class="op">&lt;-</span> <span class="va">pbmc3k</span><span class="op">$</span><span class="va">seurat_annotations</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">umap_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">UMAP_1</span>, y <span class="op">=</span> <span class="va">UMAP_2</span>, color <span class="op">=</span> <span class="va">celltypes</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="stability-dim-reduction-flexibility_files/figure-html/umap_harmony-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="clustassess-in-the-scatac-seq-data">ClustAssess in the scATAC-seq data<a class="anchor" aria-label="anchor" href="#clustassess-in-the-scatac-seq-data"></a>
</h2>
<p>In this example we will showcase the flexibility of the <code>assess_feature_stability</code> function by using the ATAC-seq data. For this example, we will use the multiome PBMC dataset from the <code>SeuratData</code> package.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stuart-lab/signac" class="external-link">Signac</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/InstallData.html" class="external-link">InstallData</a></span><span class="op">(</span><span class="st">"pbmcMultiome"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Installing package into '/tmp/RtmpHLlwfi/temp_libpath7b4c499b60ac'</span></span>
<span><span class="co">## (as 'lib' is unspecified)</span></span></code></pre>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"pbmc.atac"</span><span class="op">)</span></span></code></pre></div>
<p>As presented in the (Signac)(<a href="https://stuartlab.org/signac/articles/pbmc_vignette" class="external-link uri">https://stuartlab.org/signac/articles/pbmc_vignette</a>) package, the ATAC-seq data is usually processed using the TF-IDF normalization followed by the the calculation of the singular values. These two steps are also known as LSI (Latent Semantic Indexing).</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc.atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/RunTFIDF.html" class="external-link">RunTFIDF</a></span><span class="op">(</span><span class="va">pbmc.atac</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Performing TF-IDF normalization</span></span></code></pre>
<pre><code><span><span class="co">## Warning in RunTFIDF.default(object = GetAssayData(object = object, slot =</span></span>
<span><span class="co">## "counts"), : Some features contain 0 total counts</span></span></code></pre>
<p>Identify the highly variable peaks.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc.atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/FindTopFeatures.html" class="external-link">FindTopFeatures</a></span><span class="op">(</span><span class="va">pbmc.atac</span>, min.cutoff <span class="op">=</span> <span class="st">"q5"</span><span class="op">)</span></span>
<span><span class="va">var_peaks</span> <span class="op">&lt;-</span> <span class="va">pbmc.atac</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">ATAC</span><span class="op">@</span><span class="va">var.features</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">3000</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>To speedup the assessment, set a parallel backend with 6 cores.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">RhpcBLASctl</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RhpcBLASctl/man/RhpcBLASctl-package.html" class="external-link">blas_set_num_threads</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">ncores</span> <span class="op">&lt;-</span> <span class="fl">6</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">ncores</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">my_cluster</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster</a></span><span class="op">(</span></span>
<span>        <span class="va">ncores</span>,</span>
<span>        type <span class="op">=</span> <span class="st">"PSOCK"</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span>    <span class="fu">doParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/doParallel/man/registerDoParallel.html" class="external-link">registerDoParallel</a></span><span class="op">(</span>cl <span class="op">=</span> <span class="va">my_cluster</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Assess the stability of the dimensionality reduction by varying the number of highly variable peaks.</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">matrix_processing_function</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">dt_mtx</span>, <span class="va">actual_n_singular_values</span> <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">actual_n_singular_values</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">actual_n_singular_values</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">dt_mtx</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%/%</a></span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span>    <span class="fu">RhpcBLASctl</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RhpcBLASctl/man/RhpcBLASctl-package.html" class="external-link">blas_set_num_threads</a></span><span class="op">(</span><span class="fu">foreach</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/foreach/man/getDoParWorkers.html" class="external-link">getDoParWorkers</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">embedding</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/RunSVD.html" class="external-link">RunSVD</a></span><span class="op">(</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu">t</span><span class="op">(</span><span class="va">dt_mtx</span><span class="op">)</span>, n <span class="op">=</span> <span class="va">actual_n_singular_values</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">@</span><span class="va">cell.embeddings</span></span>
<span>    <span class="co"># remove the first component, as it does contain noise - see the Signac vignette</span></span>
<span>    <span class="va">embedding</span> <span class="op">&lt;-</span> <span class="va">embedding</span><span class="op">[</span>, <span class="fl">2</span><span class="op">:</span><span class="va">actual_n_singular_values</span><span class="op">]</span></span>
<span></span>
<span>    <span class="fu">RhpcBLASctl</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RhpcBLASctl/man/RhpcBLASctl-package.html" class="external-link">blas_set_num_threads</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">embedding</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dt_mtx</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">embedding</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"LSI_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">actual_n_singular_values</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">embedding</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">lsi_atac_feature_stability</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assess_feature_stability.html">assess_feature_stability</a></span><span class="op">(</span></span>
<span>    data_matrix <span class="op">=</span> <span class="va">pbmc.atac</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"ATAC"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">data</span>,</span>
<span>    feature_set <span class="op">=</span> <span class="va">var_peaks</span>,</span>
<span>    resolution <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">0.1</span>, to <span class="op">=</span> <span class="fl">1</span>, by <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>,</span>
<span>    steps <span class="op">=</span> <span class="va">steps</span>,</span>
<span>    n_repetitions <span class="op">=</span> <span class="va">n_repetitions</span>,</span>
<span>    feature_type <span class="op">=</span> <span class="st">"HV_peaks"</span>,</span>
<span>    graph_reduction_type <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>    matrix_processing <span class="op">=</span> <span class="va">matrix_processing_function</span>,</span>
<span>    umap_arguments <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        min_dist <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>        n_neighbors <span class="op">=</span> <span class="fl">30</span>,</span>
<span>        metric <span class="op">=</span> <span class="st">"cosine"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    ecs_thresh <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    clustering_algorithm <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: No assay specified, setting assay as RNA by default.</span></span>
<span><span class="co">## No assay specified, setting assay as RNA by default.</span></span>
<span><span class="co">## No assay specified, setting assay as RNA by default.</span></span>
<span><span class="co">## No assay specified, setting assay as RNA by default.</span></span>
<span><span class="co">## No assay specified, setting assay as RNA by default.</span></span>
<span><span class="co">## No assay specified, setting assay as RNA by default.</span></span></code></pre>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">foreach</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/foreach/man/registerDoSEQ.html" class="external-link">registerDoSEQ</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Plot the distribution of the celltypes on the UMAP embedding obtained on the top 1000 Highly Variable peaks.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">umap_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">lsi_atac_feature_stability</span><span class="op">$</span><span class="va">embedding_list</span><span class="op">$</span><span class="va">HV_peaks</span><span class="op">$</span><span class="st">"1000"</span><span class="op">)</span></span>
<span><span class="va">umap_df</span><span class="op">$</span><span class="va">celltypes</span> <span class="op">&lt;-</span> <span class="va">pbmc.atac</span><span class="op">$</span><span class="va">seurat_annotations</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">umap_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">UMAP_1</span>, y <span class="op">=</span> <span class="va">UMAP_2</span>, color <span class="op">=</span> <span class="va">celltypes</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="stability-dim-reduction-flexibility_files/figure-html/umap_atac-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.0 (2024-04-24)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Europe/Bucharest</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] pbmcMultiome.SeuratData_0.1.4 Signac_1.13.0                </span></span>
<span><span class="co">##  [3] data.table_1.15.4             ggplot2_3.5.1                </span></span>
<span><span class="co">##  [5] ClustAssess_1.0.0             Seurat_5.1.0                 </span></span>
<span><span class="co">##  [7] SeuratObject_5.0.2            sp_2.1-4                     </span></span>
<span><span class="co">##  [9] Rsamtools_2.20.0              Biostrings_2.72.1            </span></span>
<span><span class="co">## [11] XVector_0.44.0                GenomicRanges_1.56.0         </span></span>
<span><span class="co">## [13] GenomeInfoDb_1.40.1           IRanges_2.38.0               </span></span>
<span><span class="co">## [15] S4Vectors_0.42.0              BiocGenerics_0.50.0          </span></span>
<span><span class="co">## [17] harmony_1.2.1                 Rcpp_1.0.13                  </span></span>
<span><span class="co">## [19] pbmc3k.SeuratData_3.1.4       SeuratData_0.2.2.9001        </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3      jsonlite_1.8.8          magrittr_2.0.3         </span></span>
<span><span class="co">##   [4] spatstat.utils_3.1-0    farver_2.1.2            rmarkdown_2.27         </span></span>
<span><span class="co">##   [7] zlibbioc_1.50.0         fs_1.6.4                ragg_1.3.3             </span></span>
<span><span class="co">##  [10] vctrs_0.6.5             ROCR_1.0-11             spatstat.explore_3.2-7 </span></span>
<span><span class="co">##  [13] RcppRoll_0.3.0          progress_1.2.3          htmltools_0.5.8.1      </span></span>
<span><span class="co">##  [16] sass_0.4.9              sctransform_0.4.1       parallelly_1.37.1      </span></span>
<span><span class="co">##  [19] KernSmooth_2.23-22      bslib_0.7.0             htmlwidgets_1.6.4      </span></span>
<span><span class="co">##  [22] desc_1.4.3              ica_1.0-3               plyr_1.8.9             </span></span>
<span><span class="co">##  [25] plotly_4.10.4           zoo_1.8-12              cachem_1.1.0           </span></span>
<span><span class="co">##  [28] igraph_2.0.3            iterators_1.0.14        mime_0.12              </span></span>
<span><span class="co">##  [31] lifecycle_1.0.4         pkgconfig_2.0.3         Matrix_1.7-0           </span></span>
<span><span class="co">##  [34] R6_2.5.1                fastmap_1.2.0           GenomeInfoDbData_1.2.12</span></span>
<span><span class="co">##  [37] fitdistrplus_1.1-11     future_1.33.2           shiny_1.8.1.1          </span></span>
<span><span class="co">##  [40] digest_0.6.35           colorspace_2.1-1        patchwork_1.2.0        </span></span>
<span><span class="co">##  [43] tensor_1.5              RSpectra_0.16-2         irlba_2.3.5.1          </span></span>
<span><span class="co">##  [46] textshaping_0.4.0       labeling_0.4.3          progressr_0.14.0       </span></span>
<span><span class="co">##  [49] fansi_1.0.6             spatstat.sparse_3.0-3   httr_1.4.7             </span></span>
<span><span class="co">##  [52] polyclip_1.10-6         abind_1.4-5             compiler_4.4.0         </span></span>
<span><span class="co">##  [55] doParallel_1.0.17       withr_3.0.1             BiocParallel_1.38.0    </span></span>
<span><span class="co">##  [58] fastDummies_1.7.3       highr_0.11              MASS_7.3-60            </span></span>
<span><span class="co">##  [61] rappdirs_0.3.3          tools_4.4.0             lmtest_0.9-40          </span></span>
<span><span class="co">##  [64] httpuv_1.6.15           future.apply_1.11.2     goftest_1.2-3          </span></span>
<span><span class="co">##  [67] glue_1.7.0              nlme_3.1-163            promises_1.3.0         </span></span>
<span><span class="co">##  [70] grid_4.4.0              Rtsne_0.17              cluster_2.1.6          </span></span>
<span><span class="co">##  [73] reshape2_1.4.4          generics_0.1.3          gtable_0.3.5           </span></span>
<span><span class="co">##  [76] spatstat.data_3.0-4     tidyr_1.3.1             hms_1.1.3              </span></span>
<span><span class="co">##  [79] utf8_1.2.4              spatstat.geom_3.2-9     RcppAnnoy_0.0.22       </span></span>
<span><span class="co">##  [82] foreach_1.5.2           ggrepel_0.9.5           RANN_2.6.1             </span></span>
<span><span class="co">##  [85] pillar_1.9.0            stringr_1.5.1           spam_2.10-0            </span></span>
<span><span class="co">##  [88] RcppHNSW_0.6.0          later_1.3.2             splines_4.4.0          </span></span>
<span><span class="co">##  [91] dplyr_1.1.4             lattice_0.22-5          survival_3.5-8         </span></span>
<span><span class="co">##  [94] deldir_2.0-4            tidyselect_1.2.1        miniUI_0.1.1.1         </span></span>
<span><span class="co">##  [97] pbapply_1.7-2           knitr_1.47              gridExtra_2.3          </span></span>
<span><span class="co">## [100] scattermore_1.2         RhpcBLASctl_0.23-42     xfun_0.44              </span></span>
<span><span class="co">## [103] SharedObject_1.19.1     matrixStats_1.3.0       UCSC.utils_1.0.0       </span></span>
<span><span class="co">## [106] stringi_1.8.4           lazyeval_0.2.2          yaml_2.3.8             </span></span>
<span><span class="co">## [109] evaluate_1.0.0          codetools_0.2-19        tibble_3.2.1           </span></span>
<span><span class="co">## [112] cli_3.6.3               uwot_0.2.2              xtable_1.8-4           </span></span>
<span><span class="co">## [115] reticulate_1.37.0       systemfonts_1.1.0       munsell_0.5.1          </span></span>
<span><span class="co">## [118] jquerylib_0.1.4         globals_0.16.3          spatstat.random_3.2-3  </span></span>
<span><span class="co">## [121] png_0.1-8               parallel_4.4.0          pkgdown_2.1.1          </span></span>
<span><span class="co">## [124] prettyunits_1.2.0       dotCall64_1.1-1         bitops_1.0-7           </span></span>
<span><span class="co">## [127] listenv_0.9.1           viridisLite_0.4.2       scales_1.3.0           </span></span>
<span><span class="co">## [130] ggridges_0.5.6          leiden_0.4.3.1          purrr_1.0.2            </span></span>
<span><span class="co">## [133] crayon_1.5.3            rlang_1.1.4             fastmatch_1.1-4        </span></span>
<span><span class="co">## [136] cowplot_1.1.3</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Andi Munteanu, Arash Shahsavari, Miguel Larraz Lopez de Novales, Irina Mohorianu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
