<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Calculate markers - Shiny — calculate_markers_shiny • ClustAssess</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate markers - Shiny — calculate_markers_shiny"><meta property="og:description" content="Performs the Wilcoxon rank sum test to identify differentially
expressed genes between two groups of cells in the shiny context. The
method can be also used outside the shiny context, as long as the expression
matrix is stored in a h5 file."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ClustAssess</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/ClustAssess.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/comparing-soft-and-hierarchical.html">Comparing soft and hierarchical clusterings with element-centric similarity</a>
    </li>
    <li>
      <a href="../articles/stability-dim-reduction-flexibility.html">Flexibility of the dimensionality reduction assessment</a>
    </li>
    <li>
      <a href="../articles/stability-pipeline-description.html">Stability-based parameter assessment</a>
    </li>
    <li>
      <a href="../articles/stability-pipeline-shiny.html">Automatic ClustAssess pipeline. Generating the ClustAssess shiny-app</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/Core-Bioinformatics/ClustAssess/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate markers - Shiny</h1>
    <small class="dont-index">Source: <a href="https://github.com/Core-Bioinformatics/ClustAssess/blob/HEAD/R/shiny-utils.R" class="external-link"><code>R/shiny-utils.R</code></a></small>
    <div class="hidden name"><code>calculate_markers_shiny.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Performs the Wilcoxon rank sum test to identify differentially
expressed genes between two groups of cells in the shiny context. The
method can be also used outside the shiny context, as long as the expression
matrix is stored in a h5 file.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">calculate_markers_shiny</span><span class="op">(</span></span>
<span>  <span class="va">cells1</span>,</span>
<span>  <span class="va">cells2</span>,</span>
<span>  logfc_threshold <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  min_pct_threshold <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  average_expression_threshold <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  average_expression_group1_threshold <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  min_diff_pct_threshold <span class="op">=</span> <span class="op">-</span><span class="cn">Inf</span>,</span>
<span>  used_slot <span class="op">=</span> <span class="st">"data"</span>,</span>
<span>  norm_method <span class="op">=</span> <span class="st">"SCT"</span>,</span>
<span>  expression_h5_path <span class="op">=</span> <span class="st">"expression.h5"</span>,</span>
<span>  pseudocount_use <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  base <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  check_difference <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-cells-">cells1<a class="anchor" aria-label="anchor" href="#arg-cells-"></a></dt>
<dd><p>A vector of cell indices for the first group of cells.</p></dd>


<dt id="arg-cells-">cells2<a class="anchor" aria-label="anchor" href="#arg-cells-"></a></dt>
<dd><p>A vector of cell indices for the second group of cells.</p></dd>


<dt id="arg-logfc-threshold">logfc_threshold<a class="anchor" aria-label="anchor" href="#arg-logfc-threshold"></a></dt>
<dd><p>The minimum absolute log fold change to consider a
gene as differentially expressed. Defaults to <code>0</code>, meaning all genes are
taken into considereation.</p></dd>


<dt id="arg-min-pct-threshold">min_pct_threshold<a class="anchor" aria-label="anchor" href="#arg-min-pct-threshold"></a></dt>
<dd><p>The minimum fraction of cells expressing a gene
form each cell population to consider the gene as differentially expressed.
Increasing the value will speed up the function. Defaults to <code>0.1</code>.</p></dd>


<dt id="arg-average-expression-threshold">average_expression_threshold<a class="anchor" aria-label="anchor" href="#arg-average-expression-threshold"></a></dt>
<dd><p>The minimum average expression that a
gene should have in order to be considered as differentially expressed.</p></dd>


<dt id="arg-average-expression-group-threshold">average_expression_group1_threshold<a class="anchor" aria-label="anchor" href="#arg-average-expression-group-threshold"></a></dt>
<dd><p>The minimum average expression
that a gene should have in the first group of cells to be considered as
differentially expressed. Defaults to <code>0</code>.</p></dd>


<dt id="arg-min-diff-pct-threshold">min_diff_pct_threshold<a class="anchor" aria-label="anchor" href="#arg-min-diff-pct-threshold"></a></dt>
<dd><p>The minimum difference in the fraction of cells
expressing a gene between the two cell populations to consider the gene as
differentially expressed. Defaults to <code>-Inf</code>.</p></dd>


<dt id="arg-used-slot">used_slot<a class="anchor" aria-label="anchor" href="#arg-used-slot"></a></dt>
<dd><p>Parameter that provides additional information about the
expression matrix, whether it was scaled or not. The value of this parameter
impacts the calculation of the fold change. If <code>data</code>, the function will
calculates the fold change as the fraction between the log value of the
average of the expression raised to exponential for the two cell groups. If
<code>scale.data</code>, the function will calculate the fold change as the fraction
between the average of the expression values for the two cell groups.
Other options will default to calculating the fold change as the fraction
between the log value of the average of the expression values for the two
cell groups. Defaults to <code>data</code>.</p></dd>


<dt id="arg-norm-method">norm_method<a class="anchor" aria-label="anchor" href="#arg-norm-method"></a></dt>
<dd><p>The normalization method used to normalize the expression
matrix. The value of this parameter impacts the calculation of the average
expression of the genes when <code>used_slot = "data"</code>. If <code>LogNormalize</code>, the
log fold change will be calculated as described for the <code>used_slot</code>
parameter. Otherwise, the log fold change will be calculated as the fraction
between the log value of the average of the expression values for the two
cell groups. Defaults to <code>SCT</code>.</p></dd>


<dt id="arg-expression-h-path">expression_h5_path<a class="anchor" aria-label="anchor" href="#arg-expression-h-path"></a></dt>
<dd><p>The path to the h5 file containing the expression
matrix. The h5 file should contain the following fields: <code>expression_matrix</code>,
<code>rank_matrix</code>, <code>average_expression</code>, <code>genes</code>. The file path
defaults to <code>expression.h5</code>.</p></dd>


<dt id="arg-pseudocount-use">pseudocount_use<a class="anchor" aria-label="anchor" href="#arg-pseudocount-use"></a></dt>
<dd><p>The pseudocount to add to the expression values when
calculating the average expression of the genes, to avoid the 0 value for
the denominator. Defaults to <code>1</code>.</p></dd>


<dt id="arg-base">base<a class="anchor" aria-label="anchor" href="#arg-base"></a></dt>
<dd><p>The base of the logharithm. Defaults to <code>2</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Whether to print messages about the progress of the function.
Defaults to TRUE.</p></dd>


<dt id="arg-check-difference">check_difference<a class="anchor" aria-label="anchor" href="#arg-check-difference"></a></dt>
<dd><p>Whether to perform set difference between the two
cells. Defaults to TRUE.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A data frame containing the following columns:</p><ul><li><p><code>gene</code>: The gene name.</p></li>
<li><p><code>avg_log2FC</code>: The average log fold change between the two cell groups.</p></li>
<li><p><code>p_val</code>: The p-value of the Wilcoxon rank sum test.</p></li>
<li><p><code>p_val_adj</code>: The adjusted p-value of the Wilcoxon rank sum test.</p></li>
<li><p><code>pct.1</code>: The fraction of cells expressing the gene in the first cell group.</p></li>
<li><p><code>pct.2</code>: The fraction of cells expressing the gene in the second cell group.</p></li>
<li><p><code>avg_expr_group1</code>: The average expression of the gene in the first cell group.</p></li>
<li><p><code>avg_expr</code>: The average expression of the gene.</p></li>
</ul></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Andi Munteanu, Arash Shahsavari, Miguel Larraz Lopez de Novales, Irina Mohorianu.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

