<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Calculate markers - Shiny — calculate_markers_shiny • ClustAssess</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate markers - Shiny — calculate_markers_shiny"><meta name="description" content="Performs the Wilcoxon rank sum test to identify differentially
expressed genes between two groups of cells in the shiny context. The
method can be also used outside the shiny context, as long as the expression
matrix is stored in a h5 file."><meta property="og:description" content="Performs the Wilcoxon rank sum test to identify differentially
expressed genes between two groups of cells in the shiny context. The
method can be also used outside the shiny context, as long as the expression
matrix is stored in a h5 file."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ClustAssess</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/ClustAssess.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/comparing-soft-and-hierarchical.html">Comparing soft and hierarchical clusterings with element-centric similarity</a></li>
    <li><a class="dropdown-item" href="../articles/stability-dim-reduction-flexibility.html">Flexibility of the dimensionality reduction assessment</a></li>
    <li><a class="dropdown-item" href="../articles/stability-pipeline-description.html">Stability-based parameter assessment</a></li>
    <li><a class="dropdown-item" href="../articles/stability-pipeline-shiny.html">Automatic ClustAssess pipeline. Generating the ClustAssess shiny-app</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
<li class="nav-item"><a class="nav-link" href="../tutorials/index.html">ClustAssess Shiny app Tutorial</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Core-Bioinformatics/ClustAssess/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Calculate markers - Shiny</h1>
      <small class="dont-index">Source: <a href="https://github.com/Core-Bioinformatics/ClustAssess/blob/HEAD/R/shiny-utils.R" class="external-link"><code>R/shiny-utils.R</code></a></small>
      <div class="d-none name"><code>calculate_markers_shiny.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Performs the Wilcoxon rank sum test to identify differentially
expressed genes between two groups of cells in the shiny context. The
method can be also used outside the shiny context, as long as the expression
matrix is stored in a h5 file.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">calculate_markers_shiny</span><span class="op">(</span></span>
<span>  <span class="va">cells1</span>,</span>
<span>  <span class="va">cells2</span>,</span>
<span>  logfc_threshold <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  min_pct_threshold <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  average_expression_threshold <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  average_expression_group1_threshold <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  min_diff_pct_threshold <span class="op">=</span> <span class="op">-</span><span class="cn">Inf</span>,</span>
<span>  used_slot <span class="op">=</span> <span class="st">"data"</span>,</span>
<span>  norm_method <span class="op">=</span> <span class="st">"SCT"</span>,</span>
<span>  expression_h5_path <span class="op">=</span> <span class="st">"expression.h5"</span>,</span>
<span>  pseudocount_use <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  base <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  check_difference <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-cells-">cells1<a class="anchor" aria-label="anchor" href="#arg-cells-"></a></dt>
<dd><p>A vector of cell indices for the first group of cells.</p></dd>


<dt id="arg-cells-">cells2<a class="anchor" aria-label="anchor" href="#arg-cells-"></a></dt>
<dd><p>A vector of cell indices for the second group of cells.</p></dd>


<dt id="arg-logfc-threshold">logfc_threshold<a class="anchor" aria-label="anchor" href="#arg-logfc-threshold"></a></dt>
<dd><p>The minimum absolute log fold change to consider a
gene as differentially expressed. Defaults to <code>0</code>, meaning all genes are
taken into considereation.</p></dd>


<dt id="arg-min-pct-threshold">min_pct_threshold<a class="anchor" aria-label="anchor" href="#arg-min-pct-threshold"></a></dt>
<dd><p>The minimum fraction of cells expressing a gene
form each cell population to consider the gene as differentially expressed.
Increasing the value will speed up the function. Defaults to <code>0.1</code>.</p></dd>


<dt id="arg-average-expression-threshold">average_expression_threshold<a class="anchor" aria-label="anchor" href="#arg-average-expression-threshold"></a></dt>
<dd><p>The minimum average expression that a
gene should have in order to be considered as differentially expressed.</p></dd>


<dt id="arg-average-expression-group-threshold">average_expression_group1_threshold<a class="anchor" aria-label="anchor" href="#arg-average-expression-group-threshold"></a></dt>
<dd><p>The minimum average expression
that a gene should have in the first group of cells to be considered as
differentially expressed. Defaults to <code>0</code>.</p></dd>


<dt id="arg-min-diff-pct-threshold">min_diff_pct_threshold<a class="anchor" aria-label="anchor" href="#arg-min-diff-pct-threshold"></a></dt>
<dd><p>The minimum difference in the fraction of cells
expressing a gene between the two cell populations to consider the gene as
differentially expressed. Defaults to <code>-Inf</code>.</p></dd>


<dt id="arg-used-slot">used_slot<a class="anchor" aria-label="anchor" href="#arg-used-slot"></a></dt>
<dd><p>Parameter that provides additional information about the
expression matrix, whether it was scaled or not. The value of this parameter
impacts the calculation of the fold change. If <code>data</code>, the function will
calculates the fold change as the fraction between the log value of the
average of the expression raised to exponential for the two cell groups. If
<code>scale.data</code>, the function will calculate the fold change as the fraction
between the average of the expression values for the two cell groups.
Other options will default to calculating the fold change as the fraction
between the log value of the average of the expression values for the two
cell groups. Defaults to <code>data</code>.</p></dd>


<dt id="arg-norm-method">norm_method<a class="anchor" aria-label="anchor" href="#arg-norm-method"></a></dt>
<dd><p>The normalization method used to normalize the expression
matrix. The value of this parameter impacts the calculation of the average
expression of the genes when <code>used_slot = "data"</code>. If <code>LogNormalize</code>, the
log fold change will be calculated as described for the <code>used_slot</code>
parameter. Otherwise, the log fold change will be calculated as the fraction
between the log value of the average of the expression values for the two
cell groups. Defaults to <code>SCT</code>.</p></dd>


<dt id="arg-expression-h-path">expression_h5_path<a class="anchor" aria-label="anchor" href="#arg-expression-h-path"></a></dt>
<dd><p>The path to the h5 file containing the expression
matrix. The h5 file should contain the following fields: <code>expression_matrix</code>,
<code>rank_matrix</code>, <code>average_expression</code>, <code>genes</code>. The file path
defaults to <code>expression.h5</code>.</p></dd>


<dt id="arg-pseudocount-use">pseudocount_use<a class="anchor" aria-label="anchor" href="#arg-pseudocount-use"></a></dt>
<dd><p>The pseudocount to add to the expression values when
calculating the average expression of the genes, to avoid the 0 value for
the denominator. Defaults to <code>1</code>.</p></dd>


<dt id="arg-base">base<a class="anchor" aria-label="anchor" href="#arg-base"></a></dt>
<dd><p>The base of the logharithm. Defaults to <code>2</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Whether to print messages about the progress of the function.
Defaults to TRUE.</p></dd>


<dt id="arg-check-difference">check_difference<a class="anchor" aria-label="anchor" href="#arg-check-difference"></a></dt>
<dd><p>Whether to perform set difference between the two
cells. Defaults to TRUE.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data frame containing the following columns:</p><ul><li><p><code>gene</code>: The gene name.</p></li>
<li><p><code>avg_log2FC</code>: The average log fold change between the two cell groups.</p></li>
<li><p><code>p_val</code>: The p-value of the Wilcoxon rank sum test.</p></li>
<li><p><code>p_val_adj</code>: The adjusted p-value of the Wilcoxon rank sum test.</p></li>
<li><p><code>pct.1</code>: The fraction of cells expressing the gene in the first cell group.</p></li>
<li><p><code>pct.2</code>: The fraction of cells expressing the gene in the second cell group.</p></li>
<li><p><code>avg_expr_group1</code>: The average expression of the gene in the first cell group.</p></li>
<li><p><code>avg_expr</code>: The average expression of the gene.</p></li>
</ul></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Andi Munteanu, Arash Shahsavari, Miguel Larraz Lopez de Novales, Irina Mohorianu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

