<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Calculate markers — calculate_markers • ClustAssess</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate markers — calculate_markers"><meta name="description" content="Performs the Wilcoxon rank sum test to identify differentially
expressed genes between two groups of cells."><meta property="og:description" content="Performs the Wilcoxon rank sum test to identify differentially
expressed genes between two groups of cells."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ClustAssess</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/ClustAssess.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/comparing-soft-and-hierarchical.html">Comparing soft and hierarchical clusterings with element-centric similarity</a></li>
    <li><a class="dropdown-item" href="../articles/stability-dim-reduction-flexibility.html">Flexibility of the dimensionality reduction assessment</a></li>
    <li><a class="dropdown-item" href="../articles/stability-pipeline-description.html">Stability-based parameter assessment</a></li>
    <li><a class="dropdown-item" href="../articles/stability-pipeline-shiny.html">Automatic ClustAssess pipeline. Generating the ClustAssess shiny-app</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../tutorials/index.html">ClustAssess Shiny app Tutorial</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Core-Bioinformatics/ClustAssess/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Calculate markers</h1>
      <small class="dont-index">Source: <a href="https://github.com/Core-Bioinformatics/ClustAssess/blob/HEAD/R/shiny-utils.R" class="external-link"><code>R/shiny-utils.R</code></a></small>
      <div class="d-none name"><code>calculate_markers.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Performs the Wilcoxon rank sum test to identify differentially
expressed genes between two groups of cells.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">calculate_markers</span><span class="op">(</span></span>
<span>  <span class="va">expression_matrix</span>,</span>
<span>  <span class="va">cells1</span>,</span>
<span>  <span class="va">cells2</span>,</span>
<span>  logfc_threshold <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  min_pct_threshold <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  avg_expr_threshold_group1 <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  min_diff_pct_threshold <span class="op">=</span> <span class="op">-</span><span class="cn">Inf</span>,</span>
<span>  rank_matrix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  feature_names <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  used_slot <span class="op">=</span> <span class="st">"data"</span>,</span>
<span>  norm_method <span class="op">=</span> <span class="st">"SCT"</span>,</span>
<span>  pseudocount_use <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  base <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  adjust_pvals <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  check_cells_set_diff <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-expression-matrix">expression_matrix<a class="anchor" aria-label="anchor" href="#arg-expression-matrix"></a></dt>
<dd><p>A matrix of gene expression values having genes
in rows and cells in columns.</p></dd>


<dt id="arg-cells-">cells1<a class="anchor" aria-label="anchor" href="#arg-cells-"></a></dt>
<dd><p>A vector of cell indices for the first group of cells.</p></dd>


<dt id="arg-cells-">cells2<a class="anchor" aria-label="anchor" href="#arg-cells-"></a></dt>
<dd><p>A vector of cell indices for the second group of cells.</p></dd>


<dt id="arg-logfc-threshold">logfc_threshold<a class="anchor" aria-label="anchor" href="#arg-logfc-threshold"></a></dt>
<dd><p>The minimum absolute log fold change to consider a
gene as differentially expressed. Defaults to <code>0</code>, meaning all genes are
taken into considereation.</p></dd>


<dt id="arg-min-pct-threshold">min_pct_threshold<a class="anchor" aria-label="anchor" href="#arg-min-pct-threshold"></a></dt>
<dd><p>The minimum fraction of cells expressing a gene
form each cell population to consider the gene as differentially expressed.
Increasing the value will speed up the function. Defaults to <code>0.1</code>.</p></dd>


<dt id="arg-avg-expr-threshold-group-">avg_expr_threshold_group1<a class="anchor" aria-label="anchor" href="#arg-avg-expr-threshold-group-"></a></dt>
<dd><p>The minimum average expression that a gene
should have in the first group of cells to be considered as differentially
expressed. Defaults to <code>0</code>.</p></dd>


<dt id="arg-min-diff-pct-threshold">min_diff_pct_threshold<a class="anchor" aria-label="anchor" href="#arg-min-diff-pct-threshold"></a></dt>
<dd><p>The minimum difference in the fraction of cells
expressing a gene between the two cell populations to consider the gene as
differentially expressed. Defaults to <code>-Inf</code>.</p></dd>


<dt id="arg-rank-matrix">rank_matrix<a class="anchor" aria-label="anchor" href="#arg-rank-matrix"></a></dt>
<dd><p>A matrix where the cells are ranked based on their
expression levels with respect to each gene. Defaults to <code>NULL</code>, in which
case the function will calculate the rank matrix. We recommend calculating
the rank matrix beforehand and passing it to the function to speed up the
computation.</p></dd>


<dt id="arg-feature-names">feature_names<a class="anchor" aria-label="anchor" href="#arg-feature-names"></a></dt>
<dd><p>A vector of gene names. Defaults to <code>NULL</code>, in which
case the function will use the row names of the expression matrix as gene
names.</p></dd>


<dt id="arg-used-slot">used_slot<a class="anchor" aria-label="anchor" href="#arg-used-slot"></a></dt>
<dd><p>Parameter that provides additional information about the
expression matrix, whether it was scaled or not. The value of this parameter
impacts the calculation of the fold change. If <code>data</code>, the function will
calculates the fold change as the fraction between the log value of the
average of the expression raised to exponential for the two cell groups. If
<code>scale.data</code>, the function will calculate the fold change as the fraction
between the average of the expression values for the two cell groups.
Other options will default to calculating the fold change as the fraction
between the log value of the average of the expression values for the two
cell groups. Defaults to <code>data</code>.</p></dd>


<dt id="arg-norm-method">norm_method<a class="anchor" aria-label="anchor" href="#arg-norm-method"></a></dt>
<dd><p>The normalization method used to normalize the expression
matrix. The value of this parameter impacts the calculation of the average
expression of the genes when <code>used_slot = "data"</code>. If <code>LogNormalize</code>, the
log fold change will be calculated as described for the <code>used_slot</code>
parameter. Otherwise, the log fold change will be calculated as the fraction
between the log value of the average of the expression values for the two
cell groups. Defaults to <code>SCT</code>.</p></dd>


<dt id="arg-pseudocount-use">pseudocount_use<a class="anchor" aria-label="anchor" href="#arg-pseudocount-use"></a></dt>
<dd><p>The pseudocount to add to the expression values when
calculating the average expression of the genes, to avoid the 0 value for
the denominator. Defaults to <code>1</code>.</p></dd>


<dt id="arg-base">base<a class="anchor" aria-label="anchor" href="#arg-base"></a></dt>
<dd><p>The base of the logharithm. Defaults to <code>2</code>.</p></dd>


<dt id="arg-adjust-pvals">adjust_pvals<a class="anchor" aria-label="anchor" href="#arg-adjust-pvals"></a></dt>
<dd><p>A logical value indicating whether to adjust the p-values
for multiple testing using the Bonferonni method. Defaults to <code>TRUE</code>.</p></dd>


<dt id="arg-check-cells-set-diff">check_cells_set_diff<a class="anchor" aria-label="anchor" href="#arg-check-cells-set-diff"></a></dt>
<dd><p>A logical value indicating whether to check if
thw two cell groups are disjoint or not. Defaults to <code>TRUE</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data frame containing the following columns:</p><ul><li><p><code>gene</code>: The gene name.</p></li>
<li><p><code>avg_log2FC</code>: The average log fold change between the two cell groups.</p></li>
<li><p><code>p_val</code>: The p-value of the Wilcoxon rank sum test.</p></li>
<li><p><code>p_val_adj</code>: The adjusted p-value of the Wilcoxon rank sum test.</p></li>
<li><p><code>pct.1</code>: The fraction of cells expressing the gene in the first cell group.</p></li>
<li><p><code>pct.2</code>: The fraction of cells expressing the gene in the second cell group.</p></li>
<li><p><code>avg_expr_group1</code>: The average expression of the gene in the first cell group.</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2024</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># create an artificial expression matrix</span></span></span>
<span class="r-in"><span><span class="va">expr_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">100</span> <span class="op">*</span> <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">100</span> <span class="op">*</span> <span class="fl">50</span>, min <span class="op">=</span> <span class="fl">3</span>, max <span class="op">=</span> <span class="fl">4</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    ncol <span class="op">=</span> <span class="fl">200</span>, byrow <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">expr_matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">200</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">expr_matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"feature"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">calculate_markers</span><span class="op">(</span></span></span>
<span class="r-in"><span>    expression_matrix <span class="op">=</span> <span class="va">expr_matrix</span>,</span></span>
<span class="r-in"><span>    cells1 <span class="op">=</span> <span class="fl">101</span><span class="op">:</span><span class="fl">200</span>,</span></span>
<span class="r-in"><span>    cells2 <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">100</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  gene avg_log2FC pct.1 pct.2        p_val    p_val_adj</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 1   feature 1   5.324445     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 2   feature 2   5.586295     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 3   feature 3   5.273713     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 4   feature 4   5.643523     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 5   feature 5   5.706965     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 6   feature 6   5.487495     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 7   feature 7   5.487029     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 8   feature 8   5.587873     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 9   feature 9   5.578639     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 10 feature 10   5.433841     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 11 feature 11   5.616236     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 12 feature 12   5.467072     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 13 feature 13   5.444738     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 14 feature 14   5.550933     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 15 feature 15   5.464486     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 16 feature 16   5.473382     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 17 feature 17   5.536340     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 18 feature 18   5.553369     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 19 feature 19   5.550821     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 20 feature 20   5.414216     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 21 feature 21   5.339008     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 22 feature 22   5.486371     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 23 feature 23   5.514801     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 24 feature 24   5.626331     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 25 feature 25   5.520828     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 26 feature 26   5.439109     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 27 feature 27   5.510551     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 28 feature 28   5.414908     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 29 feature 29   5.536102     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 30 feature 30   5.471044     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 31 feature 31   5.712207     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 32 feature 32   5.482001     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 33 feature 33   5.396843     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 34 feature 34   5.458924     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 35 feature 35   5.466319     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 36 feature 36   5.521587     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 37 feature 37   5.585943     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 38 feature 38   5.537862     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 39 feature 39   5.531601     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 40 feature 40   5.646930     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 41 feature 41   5.322801     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 42 feature 42   5.709088     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 43 feature 43   5.784678     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 44 feature 44   5.662688     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 45 feature 45   5.541069     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 46 feature 46   5.445265     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 47 feature 47   5.515937     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 48 feature 48   5.621451     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 49 feature 49   5.516103     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 50 feature 50   5.515016     1     1 2.562144e-34 1.281072e-32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            avg_expr_group1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 1         3.497559</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 2         3.496166</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 3         3.445801</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 4         3.504912</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 5         3.577405</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 6         3.525528</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 7         3.485442</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 8         3.506795</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 9         3.476262</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 10        3.484119</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 11        3.498119</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 12        3.493887</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 13        3.471559</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 14        3.498410</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 15        3.511520</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 16        3.537302</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 17        3.482097</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 18        3.507179</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 19        3.493560</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 20        3.431861</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 21        3.465087</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 22        3.477709</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 23        3.487408</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 24        3.527158</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 25        3.513766</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 26        3.469997</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 27        3.462232</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 28        3.501975</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 29        3.491901</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 30        3.515717</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 31        3.466661</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 32        3.534487</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 33        3.483160</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 34        3.481944</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 35        3.504899</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 36        3.521041</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 37        3.508432</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 38        3.471637</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 39        3.521095</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 40        3.505936</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 41        3.480344</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 42        3.528489</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 43        3.516821</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 44        3.487346</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 45        3.502275</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 46        3.484552</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 47        3.528281</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 48        3.517302</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 49        3.513613</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature 50        3.482948</span>
<span class="r-in"><span><span class="co"># TODO should be rewritten such that you don't create new matrix objects inside</span></span></span>
<span class="r-in"><span><span class="co"># just</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Andi Munteanu, Arash Shahsavari, Miguel Larraz Lopez de Novales, Irina Mohorianu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

